<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>14 Cluster pancreatic datasets without batch correction | ANALYSIS OF SINGLE CELL RNA-SEQ DATA</title>
  <meta name="description" content="This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="14 Cluster pancreatic datasets without batch correction | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://broadinstitute.github.io/2019_scWorkshop/" />
  <meta property="og:image" content="https://broadinstitute.github.io/2019_scWorkshop/images/cover.png" />
  <meta property="og:description" content="This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook." />
  <meta name="github-repo" content="broadinstitute/2019_scWorkship" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="14 Cluster pancreatic datasets without batch correction | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  
  <meta name="twitter:description" content="This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook." />
  <meta name="twitter:image" content="https://broadinstitute.github.io/2019_scWorkshop/images/cover.png" />

<meta name="author" content="Orr Ashenberg">
<meta name="author" content="Dana Silverbush">
<meta name="author" content="Kirk Gosik">


<meta name="date" content="2019-05-30">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="correcting-batch-effects.html">
<link rel="next" href="additional-exploration-regressing-out-unwanted-covariates.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#course-overview"><i class="fa fa-check"></i><b>1.1</b> COURSE OVERVIEW</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#targeted-audience-assumed-background"><i class="fa fa-check"></i><b>1.2</b> TARGETED AUDIENCE &amp; ASSUMED BACKGROUND</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#course-format"><i class="fa fa-check"></i><b>1.3</b> COURSE FORMAT</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#getting-started"><i class="fa fa-check"></i><b>1.4</b> Getting Started</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#session-content"><i class="fa fa-check"></i><b>1.5</b> SESSION CONTENT</a><ul>
<li class="chapter" data-level="1.5.1" data-path="index.html"><a href="index.html#monday-classes-from-0930-to-1730-lunch-break-1-hr-40-min-of-total-coffee-breaks"><i class="fa fa-check"></i><b>1.5.1</b> Monday – Classes from 09:30 to 17:30 (lunch break-1 hr, 40 min of total coffee breaks)</a></li>
<li class="chapter" data-level="1.5.2" data-path="index.html"><a href="index.html#tuesday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.2</b> Tuesday – Classes from 09:30 to 17:30</a></li>
<li class="chapter" data-level="1.5.3" data-path="index.html"><a href="index.html#wednesday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.3</b> Wednesday – Classes from 09:30 to 17:30</a></li>
<li class="chapter" data-level="1.5.4" data-path="index.html"><a href="index.html#thursday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.4</b> Thursday – Classes from 09:30 to 17:30</a></li>
<li class="chapter" data-level="1.5.5" data-path="index.html"><a href="index.html#friday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.5</b> Friday – Classes from 09:30 to 17:30</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="scrna-seq-experimental-design.html"><a href="scrna-seq-experimental-design.html"><i class="fa fa-check"></i><b>2</b> scRNA-Seq Experimental Design</a></li>
<li class="chapter" data-level="3" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html"><i class="fa fa-check"></i><b>3</b> Understanding Sequencing Raw Data</a><ul>
<li class="chapter" data-level="3.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#class-environment"><i class="fa fa-check"></i><b>3.1</b> Class Environment</a><ul>
<li class="chapter" data-level="3.1.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#getting-into-aws-instance"><i class="fa fa-check"></i><b>3.1.1</b> Getting into AWS Instance</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#shell-and-unix-commands"><i class="fa fa-check"></i><b>3.2</b> Shell and Unix commands</a><ul>
<li class="chapter" data-level="3.2.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#common-linux-commands"><i class="fa fa-check"></i><b>3.2.1</b> Common Linux Commands</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#file-formats"><i class="fa fa-check"></i><b>3.3</b> File formats</a><ul>
<li class="chapter" data-level="3.3.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#view-fastq-files"><i class="fa fa-check"></i><b>3.3.1</b> View FASTQ Files</a></li>
<li class="chapter" data-level="3.3.2" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#view-bam-files"><i class="fa fa-check"></i><b>3.3.2</b> View BAM Files</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#public-data-repositories"><i class="fa fa-check"></i><b>3.4</b> Public data repositories</a><ul>
<li class="chapter" data-level="3.4.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#cellranger10x"><i class="fa fa-check"></i><b>3.4.1</b> Cellranger/10x</a></li>
<li class="chapter" data-level="3.4.2" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#geo"><i class="fa fa-check"></i><b>3.4.2</b> GEO</a></li>
<li class="chapter" data-level="3.4.3" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#single-cell-portal"><i class="fa fa-check"></i><b>3.4.3</b> Single Cell Portal</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-preprocessing.html"><a href="data-preprocessing.html"><i class="fa fa-check"></i><b>4</b> Data Preprocessing</a></li>
<li class="chapter" data-level="5" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html"><i class="fa fa-check"></i><b>5</b> Processing scRNAseq Data</a><ul>
<li class="chapter" data-level="5.1" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#goal"><i class="fa fa-check"></i><b>5.1</b> Goal</a></li>
<li class="chapter" data-level="5.2" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#further-reading"><i class="fa fa-check"></i><b>5.2</b> Further reading</a></li>
<li class="chapter" data-level="5.3" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#fastqc"><i class="fa fa-check"></i><b>5.3</b> FastQC</a><ul>
<li class="chapter" data-level="5.3.1" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#fastq-file-format"><i class="fa fa-check"></i><b>5.3.1</b> Fastq file format</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#align-the-reads"><i class="fa fa-check"></i><b>5.4</b> Align the reads</a><ul>
<li class="chapter" data-level="5.4.1" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#star-align"><i class="fa fa-check"></i><b>5.4.1</b> STAR align</a></li>
<li class="chapter" data-level="5.4.2" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#bam-file-format"><i class="fa fa-check"></i><b>5.4.2</b> Bam file format</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#visualization"><i class="fa fa-check"></i><b>5.5</b> Visualization</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="transcriptome-quantification.html"><a href="transcriptome-quantification.html"><i class="fa fa-check"></i><b>6</b> Transcriptome Quantification</a></li>
<li class="chapter" data-level="7" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html"><i class="fa fa-check"></i><b>7</b> Introduction R/Bioconductor</a><ul>
<li class="chapter" data-level="7.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#start-environment"><i class="fa fa-check"></i><b>7.1</b> Start Environment</a></li>
<li class="chapter" data-level="7.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#installing-packages"><i class="fa fa-check"></i><b>7.2</b> Installing packages</a><ul>
<li class="chapter" data-level="7.2.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#cran"><i class="fa fa-check"></i><b>7.2.1</b> CRAN</a></li>
<li class="chapter" data-level="7.2.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#github"><i class="fa fa-check"></i><b>7.2.2</b> Github</a></li>
<li class="chapter" data-level="7.2.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#bioconductor"><i class="fa fa-check"></i><b>7.2.3</b> Bioconductor</a></li>
<li class="chapter" data-level="7.2.4" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#source"><i class="fa fa-check"></i><b>7.2.4</b> Source</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#installation-instructions"><i class="fa fa-check"></i><b>7.3</b> Installation instructions:</a><ul>
<li class="chapter" data-level="7.3.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#classestypes"><i class="fa fa-check"></i><b>7.3.1</b> Classes/Types</a></li>
<li class="chapter" data-level="7.3.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#data-structures"><i class="fa fa-check"></i><b>7.3.2</b> Data structures</a></li>
<li class="chapter" data-level="7.3.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#detour-to-s3s4"><i class="fa fa-check"></i><b>7.3.3</b> Detour to S3/S4</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#more-information"><i class="fa fa-check"></i><b>7.4</b> More information</a><ul>
<li class="chapter" data-level="7.4.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#checking-for-help-for-any-function"><i class="fa fa-check"></i><b>7.4.1</b> Checking for help for any function!</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#grammer-of-graphics-ggplot2"><i class="fa fa-check"></i><b>7.5</b> Grammer of Graphics (ggplot2)</a><ul>
<li class="chapter" data-level="7.5.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#what-is-ggplot2"><i class="fa fa-check"></i><b>7.5.1</b> What is ggplot2?</a></li>
<li class="chapter" data-level="7.5.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#principles-of-ggplot2"><i class="fa fa-check"></i><b>7.5.2</b> Principles of ggplot2</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#reference"><i class="fa fa-check"></i><b>7.6</b> Reference</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="expression-qc-and-normalization.html"><a href="expression-qc-and-normalization.html"><i class="fa fa-check"></i><b>8</b> Expression QC and Normalization</a></li>
<li class="chapter" data-level="9" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html"><i class="fa fa-check"></i><b>9</b> Data Wrangling scRNAseq</a><ul>
<li class="chapter" data-level="9.1" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#goal-1"><i class="fa fa-check"></i><b>9.1</b> Goal</a></li>
<li class="chapter" data-level="9.2" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#introduction-1"><i class="fa fa-check"></i><b>9.2</b> Introduction</a><ul>
<li class="chapter" data-level="9.2.1" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#load-necessary-packages"><i class="fa fa-check"></i><b>9.2.1</b> Load necessary packages</a></li>
<li class="chapter" data-level="9.2.2" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#read-in-nsclc-counts-matrix."><i class="fa fa-check"></i><b>9.2.2</b> Read in NSCLC counts matrix.</a></li>
<li class="chapter" data-level="9.2.3" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#lets-examine-the-sparse-counts-matrix"><i class="fa fa-check"></i><b>9.2.3</b> Let’s examine the sparse counts matrix</a></li>
<li class="chapter" data-level="9.2.4" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#how-big-is-the-matrix"><i class="fa fa-check"></i><b>9.2.4</b> How big is the matrix?</a></li>
<li class="chapter" data-level="9.2.5" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#how-much-memory-does-a-sparse-matrix-take-up-relative-to-a-dense-matrix"><i class="fa fa-check"></i><b>9.2.5</b> How much memory does a sparse matrix take up relative to a dense matrix?</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#filtering-low-quality-cells"><i class="fa fa-check"></i><b>9.3</b> Filtering low-quality cells</a><ul>
<li class="chapter" data-level="9.3.1" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#look-at-the-summary-counts-for-genes-and-cells"><i class="fa fa-check"></i><b>9.3.1</b> Look at the summary counts for genes and cells</a></li>
<li class="chapter" data-level="9.3.2" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#plot-cells-ranked-by-their-number-of-detected-genes."><i class="fa fa-check"></i><b>9.3.2</b> Plot cells ranked by their number of detected genes.</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#beginning-with-seurat-httpsatijalab.orgseurat"><i class="fa fa-check"></i><b>9.4</b> Beginning with Seurat: <a href="http://satijalab.org/seurat/" class="uri">http://satijalab.org/seurat/</a></a><ul>
<li class="chapter" data-level="9.4.1" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#creating-a-seurat-object"><i class="fa fa-check"></i><b>9.4.1</b> Creating a seurat object</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#preprocessing-step-1-filter-out-low-quality-cells"><i class="fa fa-check"></i><b>9.5</b> Preprocessing step 1 : Filter out low-quality cells</a></li>
<li class="chapter" data-level="9.6" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#examine-contents-of-seurat-object"><i class="fa fa-check"></i><b>9.6</b> Examine contents of Seurat object</a><ul>
<li class="chapter" data-level="9.6.1" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#preprocessing-step-2-expression-normalization"><i class="fa fa-check"></i><b>9.6.1</b> Preprocessing step 2 : Expression normalization</a></li>
</ul></li>
<li class="chapter" data-level="9.7" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#detection-of-variable-genes-across-the-single-cells"><i class="fa fa-check"></i><b>9.7</b> Detection of variable genes across the single cells</a></li>
<li class="chapter" data-level="9.8" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#gene-set-expression-across-cells"><i class="fa fa-check"></i><b>9.8</b> Gene set expression across cells</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="identifying-cell-populations.html"><a href="identifying-cell-populations.html"><i class="fa fa-check"></i><b>10</b> Identifying Cell Populations</a><ul>
<li class="chapter" data-level="10.1" data-path="identifying-cell-populations.html"><a href="identifying-cell-populations.html#google-slides"><i class="fa fa-check"></i><b>10.1</b> Google Slides</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html"><i class="fa fa-check"></i><b>11</b> Feature Selection and Cluster Analysis</a><ul>
<li class="chapter" data-level="11.1" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#abstract"><i class="fa fa-check"></i><b>11.1</b> Abstract</a></li>
<li class="chapter" data-level="11.2" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#seurat-tutorial"><i class="fa fa-check"></i><b>11.2</b> Seurat Tutorial</a><ul>
<li class="chapter" data-level="11.2.1" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#preprocessing-steps"><i class="fa fa-check"></i><b>11.2.1</b> Preprocessing Steps</a></li>
<li class="chapter" data-level="11.2.2" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#start-of-identifying-cell-types"><i class="fa fa-check"></i><b>11.2.2</b> Start of Identifying Cell Types</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#feature-selection"><i class="fa fa-check"></i><b>11.3</b> Feature Selection</a><ul>
<li class="chapter" data-level="11.3.1" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#differential-expression-analysis"><i class="fa fa-check"></i><b>11.3.1</b> Differential Expression Analysis</a></li>
<li class="chapter" data-level="11.3.2" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#dimensionality-reduction"><i class="fa fa-check"></i><b>11.3.2</b> Dimensionality Reduction</a></li>
<li class="chapter" data-level="11.3.3" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#independent-components-analysis-ica"><i class="fa fa-check"></i><b>11.3.3</b> Independent Components Analysis (ICA)</a></li>
<li class="chapter" data-level="11.3.4" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#clustering"><i class="fa fa-check"></i><b>11.3.4</b> Clustering</a></li>
<li class="chapter" data-level="11.3.5" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#check-clusters"><i class="fa fa-check"></i><b>11.3.5</b> Check Clusters</a></li>
<li class="chapter" data-level="11.3.6" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#practice-visualizingembedding"><i class="fa fa-check"></i><b>11.3.6</b> Practice Visualizing/Embedding</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="12" data-path="batch-effects.html"><a href="batch-effects.html"><i class="fa fa-check"></i><b>12</b> Batch Effects</a></li>
<li class="chapter" data-level="13" data-path="correcting-batch-effects.html"><a href="correcting-batch-effects.html"><i class="fa fa-check"></i><b>13</b> Correcting Batch Effects</a><ul>
<li class="chapter" data-level="13.1" data-path="correcting-batch-effects.html"><a href="correcting-batch-effects.html#load-settings-and-packages"><i class="fa fa-check"></i><b>13.1</b> Load settings and packages</a></li>
<li class="chapter" data-level="13.2" data-path="correcting-batch-effects.html"><a href="correcting-batch-effects.html#preparing-the-individual-seurat-objects-for-each-pancreas-dataset-without-batch-correction"><i class="fa fa-check"></i><b>13.2</b> Preparing the individual Seurat objects for each pancreas dataset without batch correction</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="cluster-pancreatic-datasets-without-batch-correction.html"><a href="cluster-pancreatic-datasets-without-batch-correction.html"><i class="fa fa-check"></i><b>14</b> Cluster pancreatic datasets without batch correction</a><ul>
<li class="chapter" data-level="14.0.1" data-path="cluster-pancreatic-datasets-without-batch-correction.html"><a href="cluster-pancreatic-datasets-without-batch-correction.html#batch-correction-canonical-correlation-analysis-cca-using-seurat"><i class="fa fa-check"></i><b>14.0.1</b> Batch correction: canonical correlation analysis (CCA) using Seurat</a></li>
<li class="chapter" data-level="14.0.2" data-path="cluster-pancreatic-datasets-without-batch-correction.html"><a href="cluster-pancreatic-datasets-without-batch-correction.html#batch-correction-integrative-non-negative-matrix-factorization-nmf-using-liger"><i class="fa fa-check"></i><b>14.0.2</b> Batch correction: integrative non-negative matrix factorization (NMF) using LIGER</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="additional-exploration-regressing-out-unwanted-covariates.html"><a href="additional-exploration-regressing-out-unwanted-covariates.html"><i class="fa fa-check"></i><b>15</b> Additional exploration: Regressing out unwanted covariates</a></li>
<li class="chapter" data-level="16" data-path="additional-exploration-kbet.html"><a href="additional-exploration-kbet.html"><i class="fa fa-check"></i><b>16</b> Additional exploration: kBET</a></li>
<li class="chapter" data-level="17" data-path="additional-exploration-seurat-3.html"><a href="additional-exploration-seurat-3.html"><i class="fa fa-check"></i><b>17</b> Additional exploration: Seurat 3</a></li>
<li class="chapter" data-level="18" data-path="acknowledgements.html"><a href="acknowledgements.html"><i class="fa fa-check"></i><b>18</b> Acknowledgements</a></li>
<li class="chapter" data-level="19" data-path="functional-analysis.html"><a href="functional-analysis.html"><i class="fa fa-check"></i><b>19</b> Functional Analysis</a><ul>
<li class="chapter" data-level="19.1" data-path="functional-analysis.html"><a href="functional-analysis.html#google-slides-1"><i class="fa fa-check"></i><b>19.1</b> Google Slides</a></li>
<li class="chapter" data-level="19.2" data-path="functional-analysis.html"><a href="functional-analysis.html#gene-sets-and-signatures"><i class="fa fa-check"></i><b>19.2</b> Gene sets and signatures</a><ul>
<li class="chapter" data-level="19.2.1" data-path="functional-analysis.html"><a href="functional-analysis.html#cell-cycle"><i class="fa fa-check"></i><b>19.2.1</b> Cell Cycle</a></li>
</ul></li>
<li class="chapter" data-level="19.3" data-path="functional-analysis.html"><a href="functional-analysis.html#pathway-analysis"><i class="fa fa-check"></i><b>19.3</b> Pathway analysis</a></li>
<li class="chapter" data-level="19.4" data-path="functional-analysis.html"><a href="functional-analysis.html#infercnv-honeybadger"><i class="fa fa-check"></i><b>19.4</b> inferCNV / honeybadger</a><ul>
<li class="chapter" data-level="19.4.1" data-path="functional-analysis.html"><a href="functional-analysis.html#create-the-infercnv-object"><i class="fa fa-check"></i><b>19.4.1</b> Create the InferCNV Object</a></li>
<li class="chapter" data-level="19.4.2" data-path="functional-analysis.html"><a href="functional-analysis.html#filtering-genes"><i class="fa fa-check"></i><b>19.4.2</b> Filtering genes</a></li>
<li class="chapter" data-level="19.4.3" data-path="functional-analysis.html"><a href="functional-analysis.html#normalize-each-cells-counts-for-sequencing-depth"><i class="fa fa-check"></i><b>19.4.3</b> Normalize each cell’s counts for sequencing depth</a></li>
<li class="chapter" data-level="19.4.4" data-path="functional-analysis.html"><a href="functional-analysis.html#perform-anscombe-normalization"><i class="fa fa-check"></i><b>19.4.4</b> Perform Anscombe normalization</a></li>
<li class="chapter" data-level="19.4.5" data-path="functional-analysis.html"><a href="functional-analysis.html#log-transform-the-normalized-counts"><i class="fa fa-check"></i><b>19.4.5</b> Log transform the normalized counts:</a></li>
<li class="chapter" data-level="19.4.6" data-path="functional-analysis.html"><a href="functional-analysis.html#apply-maximum-bounds-to-the-expression-data-to-reduce-outlier-effects"><i class="fa fa-check"></i><b>19.4.6</b> Apply maximum bounds to the expression data to reduce outlier effects</a></li>
<li class="chapter" data-level="19.4.7" data-path="functional-analysis.html"><a href="functional-analysis.html#initial-view-before-infercnv-operations"><i class="fa fa-check"></i><b>19.4.7</b> Initial view, before inferCNV operations:</a></li>
<li class="chapter" data-level="19.4.8" data-path="functional-analysis.html"><a href="functional-analysis.html#perform-smoothing-across-chromosomes"><i class="fa fa-check"></i><b>19.4.8</b> Perform smoothing across chromosomes</a></li>
<li class="chapter" data-level="19.4.9" data-path="functional-analysis.html"><a href="functional-analysis.html#subtract-the-reference-values-from-observations-now-have-logfold-change-values"><i class="fa fa-check"></i><b>19.4.9</b> Subtract the reference values from observations, now have log(fold change) values</a></li>
<li class="chapter" data-level="19.4.10" data-path="functional-analysis.html"><a href="functional-analysis.html#invert-log-values"><i class="fa fa-check"></i><b>19.4.10</b> Invert log values</a></li>
<li class="chapter" data-level="19.4.11" data-path="functional-analysis.html"><a href="functional-analysis.html#removing-noise"><i class="fa fa-check"></i><b>19.4.11</b> Removing noise</a></li>
<li class="chapter" data-level="19.4.12" data-path="functional-analysis.html"><a href="functional-analysis.html#remove-outlier-data-points"><i class="fa fa-check"></i><b>19.4.12</b> Remove outlier data points</a></li>
<li class="chapter" data-level="19.4.13" data-path="functional-analysis.html"><a href="functional-analysis.html#find-de-genes-by-comparing-the-mutant-types-to-normal-types-basic"><i class="fa fa-check"></i><b>19.4.13</b> Find DE genes by comparing the mutant types to normal types, BASIC</a></li>
<li class="chapter" data-level="19.4.14" data-path="functional-analysis.html"><a href="functional-analysis.html#additional-information"><i class="fa fa-check"></i><b>19.4.14</b> Additional Information</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="20" data-path="pseudotime-cell-trajectories.html"><a href="pseudotime-cell-trajectories.html"><i class="fa fa-check"></i><b>20</b> Pseudotime Cell Trajectories</a><ul>
<li class="chapter" data-level="20.1" data-path="pseudotime-cell-trajectories.html"><a href="pseudotime-cell-trajectories.html#google-slides-2"><i class="fa fa-check"></i><b>20.1</b> Google Slides</a></li>
<li class="chapter" data-level="20.2" data-path="pseudotime-cell-trajectories.html"><a href="pseudotime-cell-trajectories.html#comparison-abstract"><i class="fa fa-check"></i><b>20.2</b> Comparison Abstract</a></li>
</ul></li>
<li class="chapter" data-level="21" data-path="functional-pseudotime-analysis.html"><a href="functional-pseudotime-analysis.html"><i class="fa fa-check"></i><b>21</b> Functional Pseudotime Analysis</a></li>
<li class="chapter" data-level="22" data-path="single-cell-multiomic-technologies.html"><a href="single-cell-multiomic-technologies.html"><i class="fa fa-check"></i><b>22</b> Single Cell Multiomic Technologies</a></li>
<li class="chapter" data-level="23" data-path="cite-seq-and-scatac-seq.html"><a href="cite-seq-and-scatac-seq.html"><i class="fa fa-check"></i><b>23</b> CITE-seq and scATAC-seq</a></li>
<li class="chapter" data-level="24" data-path="single-cell-resources.html"><a href="single-cell-resources.html"><i class="fa fa-check"></i><b>24</b> Single Cell Resources</a><ul>
<li class="chapter" data-level="24.1" data-path="single-cell-resources.html"><a href="single-cell-resources.html#comprehensive-list-of-single-cell-resources"><i class="fa fa-check"></i><b>24.1</b> Comprehensive list of single-cell resources</a></li>
<li class="chapter" data-level="24.2" data-path="single-cell-resources.html"><a href="single-cell-resources.html#computational-packages-for-single-cell-analysis"><i class="fa fa-check"></i><b>24.2</b> Computational packages for single-cell analysis</a></li>
<li class="chapter" data-level="24.3" data-path="single-cell-resources.html"><a href="single-cell-resources.html#elife-commentary-on-the-human-cell-atlas"><i class="fa fa-check"></i><b>24.3</b> eLife Commentary on the Human Cell Atlas</a></li>
<li class="chapter" data-level="24.4" data-path="single-cell-resources.html"><a href="single-cell-resources.html#online-courses"><i class="fa fa-check"></i><b>24.4</b> Online courses</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="cluster-pancreatic-datasets-without-batch-correction" class="section level1">
<h1><span class="header-section-number">14</span> Cluster pancreatic datasets without batch correction</h1>
<p>Let us cluster all the pancreatic islet datasets together and see whether there is a batch effect.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Merge Seurat objects by making a list of the 4 Seurat objects and using MergeMultipleSeuratObjects.</span>
<span class="co"># The documentation for this function is in the first code chunk.</span>
gcdata &lt;-<span class="st"> </span><span class="kw">MergeMultipleSeuratObjects</span>(<span class="kw">list</span>(<span class="st">&quot;celseq&quot;</span> =<span class="st"> </span>celseq, <span class="st">&quot;celseq2&quot;</span> =<span class="st"> </span>celseq2, <span class="st">&quot;fluidigmc1&quot;</span> =<span class="st"> </span>fluidigmc1, <span class="st">&quot;smartseq2&quot;</span> =<span class="st"> </span>smartseq2), <span class="dt">project =</span> <span class="st">&quot;pancreas&quot;</span>)

<span class="co"># Look at how the number of cells per gene varies across the different technologies.</span>
<span class="kw">VlnPlot</span>(gcdata, <span class="st">&quot;nGene&quot;</span>, <span class="dt">group.by =</span> <span class="st">&quot;tech&quot;</span>)

<span class="co"># The merged data must be normalized and scaled (but you only need to scale the variable genes). </span>
<span class="co"># Let us also find the variable genes again this time using all the pancreas data.</span>
gcdata &lt;-<span class="st"> </span><span class="kw">NormalizeData</span>(<span class="dt">object =</span> gcdata)
gcdata &lt;-<span class="st"> </span><span class="kw">FindVariableGenes</span>(gcdata, <span class="dt">do.plot =</span> F, <span class="dt">display.progress =</span> F)
gcdata &lt;-<span class="st"> </span><span class="kw">ScaleData</span>(gcdata, <span class="dt">genes.use =</span> gcdata<span class="op">@</span>var.genes)

<span class="co"># Do PCA on data including only the variable genes.</span>
gcdata &lt;-<span class="st"> </span><span class="kw">RunPCA</span>(gcdata, <span class="dt">pc.genes =</span> gcdata<span class="op">@</span>var.genes, <span class="dt">pcs.compute =</span> <span class="dv">30</span>, <span class="dt">do.print =</span> <span class="ot">TRUE</span>, <span class="dt">pcs.print =</span> <span class="dv">5</span>, <span class="dt">genes.print =</span> <span class="dv">5</span>)

<span class="co"># Color the PC biplot by the scRNA-seq technology. Hint: use DimPlot</span>
<span class="co"># Which technologies look similar to one another?</span>
<span class="kw">DimPlot</span>(gcdata, <span class="dt">reduction.use =</span> <span class="st">&quot;pca&quot;</span>, <span class="dt">dim.1 =</span> <span class="dv">1</span>, <span class="dt">dim.2 =</span> <span class="dv">2</span>, <span class="dt">group.by =</span> <span class="st">&quot;tech&quot;</span>)

<span class="co"># Cluster the cells using the first twenty principal components.</span>
gcdata &lt;-<span class="st"> </span><span class="kw">FindClusters</span>(gcdata, <span class="dt">reduction.type =</span> <span class="st">&quot;pca&quot;</span>, <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">20</span>, <span class="dt">print.output =</span> F, <span class="dt">save.SNN =</span> T, <span class="dt">force.recalc =</span> T, <span class="dt">random.seed =</span> <span class="dv">100</span>)

<span class="co"># Create a tSNE visualization. </span>
gcdata &lt;-<span class="st"> </span><span class="kw">RunTSNE</span>(gcdata, <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">20</span>, <span class="dt">do.fast =</span> T, <span class="dt">reduction.use =</span> <span class="st">&quot;pca&quot;</span>, <span class="dt">perplexity =</span> <span class="dv">30</span>)

<span class="co"># Visualize the Louvain clustering and the batches on the tSNE. </span>
<span class="co"># Remember, the clustering is stored in @meta.data in column res.0.8 and the technology is</span>
<span class="co"># stored in the column tech. Remember you can also use DimPlot</span>
<span class="kw">DimPlot</span>(gcdata, <span class="dt">reduction.use =</span> <span class="st">&quot;tsne&quot;</span>, <span class="dt">dim.1 =</span> <span class="dv">1</span>, <span class="dt">dim.2 =</span> <span class="dv">2</span>, <span class="dt">group.by =</span> <span class="st">&quot;res.0.8&quot;</span>)
<span class="kw">DimPlot</span>(gcdata, <span class="dt">reduction.use =</span> <span class="st">&quot;tsne&quot;</span>, <span class="dt">dim.1 =</span> <span class="dv">1</span>, <span class="dt">dim.2 =</span> <span class="dv">2</span>, <span class="dt">group.by =</span> <span class="st">&quot;tech&quot;</span>)

<span class="co"># Are you surprised by the results? Compare to your expectations from the PC biplot.</span>

<span class="co"># Adjusted rand index test for overlap between technology and cluster labelings. </span>
<span class="co"># This goes between 0 (completely dissimilar clustering) to 1 (identical clustering). </span>
<span class="co"># The adjustment corrects for chance grouping between cluster elements.</span>
<span class="co"># https://davetang.org/muse/2017/09/21/adjusted-rand-index/</span>
ari &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw">select</span>(gcdata<span class="op">@</span>meta.data, tech, res.<span class="fl">0.8</span>)
ari<span class="op">$</span>tech &lt;-<span class="st"> </span>plyr<span class="op">::</span><span class="kw">mapvalues</span>(ari<span class="op">$</span>tech, <span class="dt">from =</span> <span class="kw">c</span>(<span class="st">&quot;celseq&quot;</span>, <span class="st">&quot;celseq2&quot;</span>, <span class="st">&quot;fluidigmc1&quot;</span>, <span class="st">&quot;smartseq2&quot;</span>), <span class="dt">to =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>))
<span class="kw">adj.rand.index</span>(<span class="kw">as.numeric</span>(ari<span class="op">$</span>tech), <span class="kw">as.numeric</span>(ari<span class="op">$</span>res.<span class="fl">0.8</span>))

<span class="co"># Save current progress.</span>
<span class="co"># save(gcdata, file = Rda.path)</span>
<span class="co"># To load the data, run the following command.</span>
<span class="co"># load(Rda.path)</span></code></pre></div>
<div id="batch-correction-canonical-correlation-analysis-cca-using-seurat" class="section level3">
<h3><span class="header-section-number">14.0.1</span> Batch correction: canonical correlation analysis (CCA) using Seurat</h3>
<p>Here we use canonical correlation analysis to see to what extent it can remove potential batch effects.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># The first piece of code will identify variable genes that are highly variable in at least 2/4 datasets. We will use these variable genes in our batch correction.</span>
<span class="co"># Why would we implement such a requirement?</span>
ob.list &lt;-<span class="st"> </span><span class="kw">list</span>(celseq, celseq2, fluidigmc1, smartseq2)
genes.use &lt;-<span class="st"> </span><span class="kw">c</span>()
<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(ob.list)) {
  genes.use &lt;-<span class="st"> </span><span class="kw">c</span>(genes.use, <span class="kw">head</span>(<span class="kw">rownames</span>(ob.list[[i]]<span class="op">@</span>hvg.info), <span class="dv">1000</span>))
}
genes.use &lt;-<span class="st"> </span><span class="kw">names</span>(<span class="kw">which</span>(<span class="kw">table</span>(genes.use) <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>))
<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(ob.list)) {
  genes.use &lt;-<span class="st"> </span>genes.use[genes.use <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(ob.list[[i]]<span class="op">@</span>scale.data)]
}

<span class="co"># Run multi-set CCA on the 4 pancreatic islet datasets. </span>
<span class="co"># Use the variable genes above, and calculate 15 canonical components.</span>
gcdata.CCA &lt;-<span class="st"> </span><span class="kw">RunMultiCCA</span>(ob.list, <span class="dt">genes.use =</span> genes.use, <span class="dt">num.ccs =</span> <span class="dv">15</span>)

<span class="co"># Visualize CCA results. First plot CC1 versus CC2 with cells grouped by tech. </span>
<span class="kw">DimPlot</span>(gcdata.CCA, <span class="dt">reduction.use =</span> <span class="st">&quot;cca&quot;</span>, <span class="dt">group.by =</span> <span class="st">&quot;tech&quot;</span>, <span class="dt">pt.size =</span> <span class="fl">0.5</span>)

<span class="co"># Visualize CCA results. Second, look at a violin plot of CC1 scores with cells grouped by tech.</span>
<span class="kw">VlnPlot</span>(gcdata.CCA, <span class="dt">features.plot =</span> <span class="kw">c</span>(<span class="st">&quot;CC1&quot;</span>, <span class="st">&quot;CC2&quot;</span>), <span class="dt">group.by =</span> <span class="st">&quot;tech&quot;</span>)

<span class="co"># We can also look at the genes important in the first few canonical components.</span>
<span class="kw">DimHeatmap</span>(<span class="dt">object =</span> gcdata.CCA, <span class="dt">reduction.type =</span> <span class="st">&quot;cca&quot;</span>, <span class="dt">cells.use =</span> <span class="dv">500</span>, 
    <span class="dt">dim.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">9</span>, <span class="dt">do.balanced =</span> <span class="ot">TRUE</span>)

<span class="co"># Read the documentation for MetageneBicorPlot and </span>
<span class="co"># also read https://en.wikipedia.org/wiki/Biweight_midcorrelation</span>
<span class="co"># Use this function to determine how many canonical components to select.</span>
<span class="co"># Remember to use the appropriate grouping variable.</span>
<span class="co"># Try evaluating the first 15 canonical components.</span>
<span class="kw">MetageneBicorPlot</span>(gcdata.CCA, <span class="dt">grouping.var =</span> <span class="st">&quot;tech&quot;</span>, <span class="dt">dims.eval =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">15</span>)

<span class="co"># Based on the previous heatmaps and the midcorrelation plot, how many CCs will you select?</span>

<span class="co"># Next we can align the CCA subspaces across the 4 datasets. This will generate a new </span>
<span class="co"># dimensional reduction called cca.aligned. The cells from all the datasets can then</span>
<span class="co"># be clustered in this new space.</span>
<span class="co"># Use AlignSubspace, specifying which variable to group by and the number of CCs to align</span>
gcdata.CCA &lt;-<span class="st"> </span><span class="kw">AlignSubspace</span>(gcdata.CCA, <span class="dt">grouping.var =</span> <span class="st">&quot;tech&quot;</span>, <span class="dt">dims.align =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">12</span>)

<span class="co"># Visualize the distribution of the aligned canonical correlation vectors (ACC1, ACC2) </span>
<span class="co"># using VlnPlot.</span>
<span class="kw">VlnPlot</span>(gcdata.CCA, <span class="dt">features.plot =</span> <span class="st">&quot;ACC1&quot;</span>, <span class="dt">group.by =</span> <span class="st">&quot;tech&quot;</span>)
<span class="kw">VlnPlot</span>(gcdata.CCA, <span class="dt">features.plot =</span> <span class="st">&quot;ACC2&quot;</span>, <span class="dt">group.by =</span> <span class="st">&quot;orig.ident&quot;</span>, <span class="dt">do.return =</span> <span class="ot">TRUE</span>)
<span class="co"># How does this compare to the distributions of the canonical components (CC1, CC2)?</span>

<span class="co"># Clustering. Choose the dimensional reduction type to use and the number of aligned </span>
<span class="co"># canonical correlation vectors to use.</span>
gcdata.CCA &lt;-<span class="st"> </span><span class="kw">FindClusters</span>(gcdata.CCA, <span class="dt">reduction.type =</span> <span class="st">&quot;cca.aligned&quot;</span>, <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">12</span>, <span class="dt">save.SNN =</span> T, <span class="dt">random.seed =</span> <span class="dv">100</span>)

<span class="co"># tSNE. Choose the dimensional reduction type to use and the number of aligned </span>
<span class="co"># canonical correlation vectors to use.</span>
gcdata.CCA &lt;-<span class="st"> </span><span class="kw">RunTSNE</span>(gcdata.CCA, <span class="dt">reduction.use =</span> <span class="st">&quot;cca.aligned&quot;</span>, <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">12</span>, <span class="dt">do.fast =</span> <span class="ot">TRUE</span>, <span class="dt">seed.use =</span> <span class="dv">1</span>)
<span class="co"># gcdata.CCA &lt;- RunTNSE()</span>

<span class="co"># Visualize the Louvain clustering and the batches on the tSNE. </span>
<span class="co"># Remember, the clustering is stored in @meta.data in column res.0.8 and the technology is</span>
<span class="co"># stored in the column tech. Remember you can also use DimPlot</span>
p1 &lt;-<span class="st"> </span><span class="kw">DimPlot</span>(gcdata.CCA, <span class="dt">reduction.use =</span> <span class="st">&quot;tsne&quot;</span>, <span class="dt">dim.1 =</span> <span class="dv">1</span>, <span class="dt">dim.2 =</span> <span class="dv">2</span>, <span class="dt">group.by =</span> <span class="st">&quot;res.0.8&quot;</span>, <span class="dt">do.return =</span> T)
p2 &lt;-<span class="st"> </span><span class="kw">DimPlot</span>(gcdata.CCA, <span class="dt">reduction.use =</span> <span class="st">&quot;tsne&quot;</span>, <span class="dt">dim.1 =</span> <span class="dv">1</span>, <span class="dt">dim.2 =</span> <span class="dv">2</span>, <span class="dt">group.by =</span> <span class="st">&quot;tech&quot;</span>, <span class="dt">do.return =</span> T)
<span class="kw">plot_grid</span>(p1, p2)

<span class="co"># Let&#39;s look to see how the adjusted rand index changed compared to using no batch correction.</span>
ari &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw">select</span>(gcdata.CCA<span class="op">@</span>meta.data, tech, res.<span class="fl">0.8</span>)
ari<span class="op">$</span>tech &lt;-<span class="st"> </span>plyr<span class="op">::</span><span class="kw">mapvalues</span>(ari<span class="op">$</span>tech, <span class="dt">from =</span> <span class="kw">c</span>(<span class="st">&quot;celseq&quot;</span>, <span class="st">&quot;celseq2&quot;</span>, <span class="st">&quot;fluidigmc1&quot;</span>, <span class="st">&quot;smartseq2&quot;</span>), <span class="dt">to =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>))
<span class="kw">adj.rand.index</span>(<span class="kw">as.numeric</span>(ari<span class="op">$</span>tech), <span class="kw">as.numeric</span>(ari<span class="op">$</span>res.<span class="fl">0.8</span>))

<span class="co"># We can also identify conserved marker genes across the batches. Differential gene expression is</span>
<span class="co"># done across each batch, and the p-values are combined.</span>
markers &lt;-<span class="st"> </span><span class="kw">FindConservedMarkers</span>(gcdata.CCA, <span class="dt">ident.1 =</span> <span class="dv">0</span>, <span class="dt">grouping.var =</span> <span class="st">&quot;tech&quot;</span>, <span class="dt">print.bar =</span> T)
<span class="kw">head</span>(markers)

<span class="co"># Visualize the expression of the first 5 marker genes on tSNE across the different batches</span>
<span class="co"># using FeatureHeatmap.</span>
<span class="kw">FeatureHeatmap</span>(gcdata.CCA, <span class="dt">features.plot =</span> <span class="kw">rownames</span>(markers)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>], <span class="dt">group.by =</span> <span class="st">&quot;tech&quot;</span>, <span class="dt">pt.size =</span> <span class="fl">0.25</span>, <span class="dt">key.position =</span> <span class="st">&quot;top&quot;</span>, <span class="dt">max.exp =</span> <span class="dv">3</span>)

<span class="co"># Save current progress.</span>
<span class="co"># save(gcdata.CCA, file = Rda.CCA.path)</span>
<span class="co"># To load the data, run the following command.</span>
<span class="co"># load(Rda.CCA.path)</span></code></pre></div>
</div>
<div id="batch-correction-integrative-non-negative-matrix-factorization-nmf-using-liger" class="section level3">
<h3><span class="header-section-number">14.0.2</span> Batch correction: integrative non-negative matrix factorization (NMF) using LIGER</h3>
<p>Here we use integrative non-negative matrix factorization to see to what extent it can remove potential batch effects.</p>
<p>The important parameters in the batch correction are the number of factors (k), the penalty parameter (lambda), and the clustering resolution. The number of factors sets the number of factors (consisting of shared and dataset-specific factors) used in factorizing the matrix. The penalty parameter sets the balance between factors shared across the batches and factors specific to the individual batches. The default setting of lambda=5.0 is usually used by the Macosko lab. Resolution=1.0 is used in the Louvain clustering of the shared neighbor factors that have been quantile normalized.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># The first piece of code will identify variable genes that are highly variable in at least 2/4 datasets. We will use these variable genes in our batch correction.</span>
ob.list &lt;-<span class="st"> </span><span class="kw">list</span>(celseq, celseq2, fluidigmc1, smartseq2)
genes.use &lt;-<span class="st"> </span><span class="kw">c</span>()
<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(ob.list)) {
  genes.use &lt;-<span class="st"> </span><span class="kw">c</span>(genes.use, <span class="kw">head</span>(<span class="kw">rownames</span>(ob.list[[i]]<span class="op">@</span>hvg.info), <span class="dv">1000</span>))
}
genes.use &lt;-<span class="st"> </span><span class="kw">names</span>(<span class="kw">which</span>(<span class="kw">table</span>(genes.use) <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>))
<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(ob.list)) {
  genes.use &lt;-<span class="st"> </span>genes.use[genes.use <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(ob.list[[i]]<span class="op">@</span>scale.data)]
}

<span class="co"># Next we create a LIGER object with raw counts data from each batch.</span>
ob.list &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="st">&quot;celseq&quot;</span> =<span class="st"> </span>celseq.data, <span class="st">&quot;celseq2&quot;</span> =<span class="st"> </span>celseq2.data, <span class="st">&quot;fluidigmc1&quot;</span> =<span class="st"> </span>fluidigmc1.data, <span class="st">&quot;smartseq2&quot;</span> =<span class="st"> </span>smartseq2.data)
data.liger &lt;-<span class="st"> </span><span class="kw">createLiger</span>(ob.list) 

<span class="co"># Normalize gene expression for each batch.</span>
data.liger &lt;-<span class="st"> </span>liger<span class="op">::</span><span class="kw">normalize</span>(data.liger)

<span class="co"># For variable gene selection, we can either use those we identified in the earlier CCA </span>
<span class="co"># batch correction analysis (genes.use) or we can use the LIGER function selectGenes().</span>
<span class="co"># data.liger &lt;- selectGenes(data.liger, var.thresh = 0.1)</span>
data.liger<span class="op">@</span>var.genes &lt;-<span class="st"> </span>genes.use

<span class="co"># Print out the number of variable genes for LIGER analysis.</span>
<span class="kw">print</span>(<span class="kw">length</span>(data.liger<span class="op">@</span>var.genes))

<span class="co"># Scale the gene expression across the datasets. </span>
<span class="co"># Why does LIGER not center the data? Hint, think about the use of </span>
<span class="co"># non-negative matrix factorization and the constraints that this imposes.</span>
data.liger &lt;-<span class="st"> </span><span class="kw">scaleNotCenter</span>(data.liger)

<span class="co"># These two steps take 10-20 min. Only run them if you finish with the rest of the code.</span>
<span class="co"># Use the `suggestK` function to determine the appropriate number of factors to use.</span>
<span class="co"># Use the `suggestLambda` function to find the smallest lambda for which the alignment metric stabilizes.</span>
<span class="co"># suggestK(data.liger)</span>
<span class="co"># suggestLambda(ligerex, k = 20)</span>

<span class="co"># Use alternating least squares (ALS) to factorize the matrix.</span>
k.suggest &lt;-<span class="st"> </span><span class="dv">20</span>  <span class="co"># with this line, we do not use the suggested k by suggestK()</span>
data.liger &lt;-<span class="st"> </span><span class="kw">optimizeALS</span>(data.liger, <span class="dt">k =</span> k.suggest, <span class="dt">rand.seed =</span> <span class="dv">1</span>) 

<span class="co"># What do matrices H, V, and W represent, and what are their dimensions?</span>
<span class="kw">dim</span>(data.liger<span class="op">@</span>H<span class="op">$</span>celseq)
<span class="kw">dim</span>(data.liger<span class="op">@</span>V<span class="op">$</span>celseq)
<span class="kw">dim</span>(data.liger<span class="op">@</span>W)

<span class="co"># Let&#39;s see what the integrated data looks like mapped onto a tSNE visualization.</span>
data.liger &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(data.liger, <span class="dt">use.raw =</span> T)
p &lt;-<span class="st"> </span><span class="kw">plotByDatasetAndCluster</span>(data.liger, <span class="dt">return.plots =</span> T)
<span class="kw">print</span>(p[[<span class="dv">1</span>]])  <span class="co"># plot by dataset</span>

<span class="co"># Next, do clustering of cells in shared nearest factor space, and then quantile alignment.</span>
data.liger &lt;-<span class="st"> </span><span class="kw">quantileAlignSNF</span>(data.liger, <span class="dt">resolution =</span> <span class="fl">1.0</span>) <span class="co"># SNF clustering and quantile alignment</span>

<span class="co"># What are the dimensions of H.norm. What does this represent? </span>
<span class="kw">dim</span>(data.liger<span class="op">@</span>H.norm)

<span class="co"># Visualize liger batch correction results.</span>
data.liger &lt;-<span class="st"> </span><span class="kw">runTSNE</span>(data.liger)
p &lt;-<span class="st"> </span><span class="kw">plotByDatasetAndCluster</span>(data.liger, <span class="dt">return.plots =</span> T) 
<span class="kw">print</span>(p[[<span class="dv">1</span>]])  <span class="co"># plot by dataset</span>
<span class="kw">plot_grid</span>(p[[<span class="dv">1</span>]], p[[<span class="dv">2</span>]])

<span class="co"># Let&#39;s look to see how the adjusted rand index changed compared to using no batch correction.</span>
tech &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(data.liger<span class="op">@</span>H), <span class="cf">function</span>(x) { 
  <span class="kw">rep</span>(<span class="kw">names</span>(data.liger<span class="op">@</span>H)[x], <span class="kw">nrow</span>(data.liger<span class="op">@</span>H[[x]]))}))
clusters &lt;-<span class="st"> </span>data.liger<span class="op">@</span>alignment.clusters
ari &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="st">&quot;tech&quot;</span> =<span class="st"> </span>tech, <span class="st">&quot;clusters&quot;</span> =<span class="st"> </span>clusters)
ari<span class="op">$</span>tech &lt;-<span class="st"> </span>plyr<span class="op">::</span><span class="kw">mapvalues</span>(ari<span class="op">$</span>tech, <span class="dt">from =</span> <span class="kw">c</span>(<span class="st">&quot;celseq&quot;</span>, <span class="st">&quot;celseq2&quot;</span>, <span class="st">&quot;fluidigmc1&quot;</span>, <span class="st">&quot;smartseq2&quot;</span>), <span class="dt">to =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>))
<span class="kw">adj.rand.index</span>(<span class="kw">as.numeric</span>(ari<span class="op">$</span>tech), <span class="kw">as.numeric</span>(ari<span class="op">$</span>clusters))

<span class="co"># Look at proportion of each batch in each cluster, and look at factor loadings across clusters</span>
<span class="kw">plotClusterProportions</span>(data.liger)
<span class="kw">plotClusterFactors</span>(data.liger, <span class="dt">use.aligned =</span> T)

<span class="co"># Look at genes that are specific to a dataset and shared across datasets.</span>
<span class="co"># Use the plotWordClouds function and choose 2 datasets.</span>
<span class="kw">pdf</span>(<span class="kw">paste0</span>(mydir, <span class="st">&quot;word_clouds.pdf&quot;</span>))
<span class="kw">plotWordClouds</span>(data.liger, <span class="dt">dataset1 =</span> <span class="st">&quot;celseq2&quot;</span>, <span class="dt">dataset2 =</span> <span class="st">&quot;smartseq2&quot;</span>)
<span class="kw">dev.off</span>()

<span class="co"># Look at factor loadings for each cell using plotFactors. </span>
<span class="kw">pdf</span>(<span class="kw">paste0</span>(mydir, <span class="st">&quot;plot_factors.pdf&quot;</span>))
<span class="kw">plotFactors</span>(data.liger)
<span class="kw">dev.off</span>()

<span class="co"># Identify shared and batch-specific marker genes from liger factorization.</span>
<span class="co"># Use the getFactorMarkers function and choose 2 datasets.</span>
<span class="co"># Then plot some genes of interest using plotGene and plotGeneViolin.</span>
markers &lt;-<span class="st"> </span><span class="kw">getFactorMarkers</span>(data.liger, <span class="dt">dataset1 =</span> <span class="st">&quot;celseq2&quot;</span>, <span class="dt">dataset2 =</span> <span class="st">&quot;smartseq2&quot;</span>, <span class="dt">num.genes =</span> <span class="dv">10</span>)
<span class="kw">plotGene</span>(data.liger, <span class="dt">gene =</span> <span class="st">&quot;INS&quot;</span>)
<span class="kw">plotGeneViolin</span>(data.liger, <span class="dt">gene =</span> <span class="st">&quot;INS&quot;</span>)

<span class="co"># Save current progress.</span>
<span class="co"># save(data.liger, file = Rda.liger.path)</span>
<span class="co"># To load the data, run the following command.</span>
<span class="co"># load(Rda.liger.path)</span></code></pre></div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="correcting-batch-effects.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="additional-exploration-regressing-out-unwanted-covariates.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["2019_scWorkshop.pdf", "2019_scWorkshop.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
