<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>11 Feature Selection and Cluster Analysis | ANALYSIS OF SINGLE CELL RNA-SEQ DATA</title>
  <meta name="description" content="This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="11 Feature Selection and Cluster Analysis | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://broadinstitute.github.io/2019_scWorkshop/" />
  <meta property="og:image" content="https://broadinstitute.github.io/2019_scWorkshop/images/cover.png" />
  <meta property="og:description" content="This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook." />
  <meta name="github-repo" content="broadinstitute/2019_scWorkship" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="11 Feature Selection and Cluster Analysis | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  
  <meta name="twitter:description" content="This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook." />
  <meta name="twitter:image" content="https://broadinstitute.github.io/2019_scWorkshop/images/cover.png" />

<meta name="author" content="Orr Ashenberg">
<meta name="author" content="Dana Silverbush">
<meta name="author" content="Kirk Gosik">


<meta name="date" content="2019-05-30">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="identifying-cell-populations.html">
<link rel="next" href="batch-effects.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#course-overview"><i class="fa fa-check"></i><b>1.1</b> COURSE OVERVIEW</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#targeted-audience-assumed-background"><i class="fa fa-check"></i><b>1.2</b> TARGETED AUDIENCE &amp; ASSUMED BACKGROUND</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#course-format"><i class="fa fa-check"></i><b>1.3</b> COURSE FORMAT</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#getting-started"><i class="fa fa-check"></i><b>1.4</b> Getting Started</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#session-content"><i class="fa fa-check"></i><b>1.5</b> SESSION CONTENT</a><ul>
<li class="chapter" data-level="1.5.1" data-path="index.html"><a href="index.html#monday-classes-from-0930-to-1730-lunch-break-1-hr-40-min-of-total-coffee-breaks"><i class="fa fa-check"></i><b>1.5.1</b> Monday – Classes from 09:30 to 17:30 (lunch break-1 hr, 40 min of total coffee breaks)</a></li>
<li class="chapter" data-level="1.5.2" data-path="index.html"><a href="index.html#tuesday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.2</b> Tuesday – Classes from 09:30 to 17:30</a></li>
<li class="chapter" data-level="1.5.3" data-path="index.html"><a href="index.html#wednesday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.3</b> Wednesday – Classes from 09:30 to 17:30</a></li>
<li class="chapter" data-level="1.5.4" data-path="index.html"><a href="index.html#thursday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.4</b> Thursday – Classes from 09:30 to 17:30</a></li>
<li class="chapter" data-level="1.5.5" data-path="index.html"><a href="index.html#friday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.5</b> Friday – Classes from 09:30 to 17:30</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="scrna-seq-experimental-design.html"><a href="scrna-seq-experimental-design.html"><i class="fa fa-check"></i><b>2</b> scRNA-Seq Experimental Design</a></li>
<li class="chapter" data-level="3" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html"><i class="fa fa-check"></i><b>3</b> Understanding Sequencing Raw Data</a><ul>
<li class="chapter" data-level="3.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#class-environment"><i class="fa fa-check"></i><b>3.1</b> Class Environment</a><ul>
<li class="chapter" data-level="3.1.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#getting-into-aws-instance"><i class="fa fa-check"></i><b>3.1.1</b> Getting into AWS Instance</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#shell-and-unix-commands"><i class="fa fa-check"></i><b>3.2</b> Shell and Unix commands</a><ul>
<li class="chapter" data-level="3.2.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#common-linux-commands"><i class="fa fa-check"></i><b>3.2.1</b> Common Linux Commands</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#file-formats"><i class="fa fa-check"></i><b>3.3</b> File formats</a><ul>
<li class="chapter" data-level="3.3.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#view-fastq-files"><i class="fa fa-check"></i><b>3.3.1</b> View FASTQ Files</a></li>
<li class="chapter" data-level="3.3.2" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#view-bam-files"><i class="fa fa-check"></i><b>3.3.2</b> View BAM Files</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#public-data-repositories"><i class="fa fa-check"></i><b>3.4</b> Public data repositories</a><ul>
<li class="chapter" data-level="3.4.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#cellranger10x"><i class="fa fa-check"></i><b>3.4.1</b> Cellranger/10x</a></li>
<li class="chapter" data-level="3.4.2" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#geo"><i class="fa fa-check"></i><b>3.4.2</b> GEO</a></li>
<li class="chapter" data-level="3.4.3" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#single-cell-portal"><i class="fa fa-check"></i><b>3.4.3</b> Single Cell Portal</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-preprocessing.html"><a href="data-preprocessing.html"><i class="fa fa-check"></i><b>4</b> Data Preprocessing</a></li>
<li class="chapter" data-level="5" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html"><i class="fa fa-check"></i><b>5</b> Processing scRNAseq Data</a><ul>
<li class="chapter" data-level="5.1" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#goal"><i class="fa fa-check"></i><b>5.1</b> Goal</a></li>
<li class="chapter" data-level="5.2" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#further-reading"><i class="fa fa-check"></i><b>5.2</b> Further reading</a></li>
<li class="chapter" data-level="5.3" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#fastqc"><i class="fa fa-check"></i><b>5.3</b> FastQC</a><ul>
<li class="chapter" data-level="5.3.1" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#fastq-file-format"><i class="fa fa-check"></i><b>5.3.1</b> Fastq file format</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#align-the-reads"><i class="fa fa-check"></i><b>5.4</b> Align the reads</a><ul>
<li class="chapter" data-level="5.4.1" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#star-align"><i class="fa fa-check"></i><b>5.4.1</b> STAR align</a></li>
<li class="chapter" data-level="5.4.2" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#bam-file-format"><i class="fa fa-check"></i><b>5.4.2</b> Bam file format</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#visualization"><i class="fa fa-check"></i><b>5.5</b> Visualization</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="transcriptome-quantification.html"><a href="transcriptome-quantification.html"><i class="fa fa-check"></i><b>6</b> Transcriptome Quantification</a></li>
<li class="chapter" data-level="7" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html"><i class="fa fa-check"></i><b>7</b> Introduction R/Bioconductor</a><ul>
<li class="chapter" data-level="7.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#start-environment"><i class="fa fa-check"></i><b>7.1</b> Start Environment</a></li>
<li class="chapter" data-level="7.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#installing-packages"><i class="fa fa-check"></i><b>7.2</b> Installing packages</a><ul>
<li class="chapter" data-level="7.2.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#cran"><i class="fa fa-check"></i><b>7.2.1</b> CRAN</a></li>
<li class="chapter" data-level="7.2.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#github"><i class="fa fa-check"></i><b>7.2.2</b> Github</a></li>
<li class="chapter" data-level="7.2.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#bioconductor"><i class="fa fa-check"></i><b>7.2.3</b> Bioconductor</a></li>
<li class="chapter" data-level="7.2.4" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#source"><i class="fa fa-check"></i><b>7.2.4</b> Source</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#installation-instructions"><i class="fa fa-check"></i><b>7.3</b> Installation instructions:</a><ul>
<li class="chapter" data-level="7.3.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#classestypes"><i class="fa fa-check"></i><b>7.3.1</b> Classes/Types</a></li>
<li class="chapter" data-level="7.3.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#data-structures"><i class="fa fa-check"></i><b>7.3.2</b> Data structures</a></li>
<li class="chapter" data-level="7.3.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#detour-to-s3s4"><i class="fa fa-check"></i><b>7.3.3</b> Detour to S3/S4</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#more-information"><i class="fa fa-check"></i><b>7.4</b> More information</a><ul>
<li class="chapter" data-level="7.4.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#checking-for-help-for-any-function"><i class="fa fa-check"></i><b>7.4.1</b> Checking for help for any function!</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#grammer-of-graphics-ggplot2"><i class="fa fa-check"></i><b>7.5</b> Grammer of Graphics (ggplot2)</a><ul>
<li class="chapter" data-level="7.5.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#what-is-ggplot2"><i class="fa fa-check"></i><b>7.5.1</b> What is ggplot2?</a></li>
<li class="chapter" data-level="7.5.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#principles-of-ggplot2"><i class="fa fa-check"></i><b>7.5.2</b> Principles of ggplot2</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#reference"><i class="fa fa-check"></i><b>7.6</b> Reference</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="expression-qc-and-normalization.html"><a href="expression-qc-and-normalization.html"><i class="fa fa-check"></i><b>8</b> Expression QC and Normalization</a></li>
<li class="chapter" data-level="9" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html"><i class="fa fa-check"></i><b>9</b> Data Wrangling scRNAseq</a><ul>
<li class="chapter" data-level="9.1" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#goal-1"><i class="fa fa-check"></i><b>9.1</b> Goal</a></li>
<li class="chapter" data-level="9.2" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#introduction-1"><i class="fa fa-check"></i><b>9.2</b> Introduction</a><ul>
<li class="chapter" data-level="9.2.1" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#load-necessary-packages"><i class="fa fa-check"></i><b>9.2.1</b> Load necessary packages</a></li>
<li class="chapter" data-level="9.2.2" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#read-in-nsclc-counts-matrix."><i class="fa fa-check"></i><b>9.2.2</b> Read in NSCLC counts matrix.</a></li>
<li class="chapter" data-level="9.2.3" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#lets-examine-the-sparse-counts-matrix"><i class="fa fa-check"></i><b>9.2.3</b> Let’s examine the sparse counts matrix</a></li>
<li class="chapter" data-level="9.2.4" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#how-big-is-the-matrix"><i class="fa fa-check"></i><b>9.2.4</b> How big is the matrix?</a></li>
<li class="chapter" data-level="9.2.5" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#how-much-memory-does-a-sparse-matrix-take-up-relative-to-a-dense-matrix"><i class="fa fa-check"></i><b>9.2.5</b> How much memory does a sparse matrix take up relative to a dense matrix?</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#filtering-low-quality-cells"><i class="fa fa-check"></i><b>9.3</b> Filtering low-quality cells</a><ul>
<li class="chapter" data-level="9.3.1" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#look-at-the-summary-counts-for-genes-and-cells"><i class="fa fa-check"></i><b>9.3.1</b> Look at the summary counts for genes and cells</a></li>
<li class="chapter" data-level="9.3.2" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#plot-cells-ranked-by-their-number-of-detected-genes."><i class="fa fa-check"></i><b>9.3.2</b> Plot cells ranked by their number of detected genes.</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#beginning-with-seurat-httpsatijalab.orgseurat"><i class="fa fa-check"></i><b>9.4</b> Beginning with Seurat: <a href="http://satijalab.org/seurat/" class="uri">http://satijalab.org/seurat/</a></a><ul>
<li class="chapter" data-level="9.4.1" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#creating-a-seurat-object"><i class="fa fa-check"></i><b>9.4.1</b> Creating a seurat object</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#preprocessing-step-1-filter-out-low-quality-cells"><i class="fa fa-check"></i><b>9.5</b> Preprocessing step 1 : Filter out low-quality cells</a></li>
<li class="chapter" data-level="9.6" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#examine-contents-of-seurat-object"><i class="fa fa-check"></i><b>9.6</b> Examine contents of Seurat object</a><ul>
<li class="chapter" data-level="9.6.1" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#preprocessing-step-2-expression-normalization"><i class="fa fa-check"></i><b>9.6.1</b> Preprocessing step 2 : Expression normalization</a></li>
</ul></li>
<li class="chapter" data-level="9.7" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#detection-of-variable-genes-across-the-single-cells"><i class="fa fa-check"></i><b>9.7</b> Detection of variable genes across the single cells</a></li>
<li class="chapter" data-level="9.8" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#gene-set-expression-across-cells"><i class="fa fa-check"></i><b>9.8</b> Gene set expression across cells</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="identifying-cell-populations.html"><a href="identifying-cell-populations.html"><i class="fa fa-check"></i><b>10</b> Identifying Cell Populations</a><ul>
<li class="chapter" data-level="10.1" data-path="identifying-cell-populations.html"><a href="identifying-cell-populations.html#google-slides"><i class="fa fa-check"></i><b>10.1</b> Google Slides</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html"><i class="fa fa-check"></i><b>11</b> Feature Selection and Cluster Analysis</a><ul>
<li class="chapter" data-level="11.1" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#abstract"><i class="fa fa-check"></i><b>11.1</b> Abstract</a></li>
<li class="chapter" data-level="11.2" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#seurat-tutorial"><i class="fa fa-check"></i><b>11.2</b> Seurat Tutorial</a><ul>
<li class="chapter" data-level="11.2.1" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#preprocessing-steps"><i class="fa fa-check"></i><b>11.2.1</b> Preprocessing Steps</a></li>
<li class="chapter" data-level="11.2.2" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#start-of-identifying-cell-types"><i class="fa fa-check"></i><b>11.2.2</b> Start of Identifying Cell Types</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#feature-selection"><i class="fa fa-check"></i><b>11.3</b> Feature Selection</a><ul>
<li class="chapter" data-level="11.3.1" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#differential-expression-analysis"><i class="fa fa-check"></i><b>11.3.1</b> Differential Expression Analysis</a></li>
<li class="chapter" data-level="11.3.2" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#dimensionality-reduction"><i class="fa fa-check"></i><b>11.3.2</b> Dimensionality Reduction</a></li>
<li class="chapter" data-level="11.3.3" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#independent-components-analysis-ica"><i class="fa fa-check"></i><b>11.3.3</b> Independent Components Analysis (ICA)</a></li>
<li class="chapter" data-level="11.3.4" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#clustering"><i class="fa fa-check"></i><b>11.3.4</b> Clustering</a></li>
<li class="chapter" data-level="11.3.5" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#check-clusters"><i class="fa fa-check"></i><b>11.3.5</b> Check Clusters</a></li>
<li class="chapter" data-level="11.3.6" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#practice-visualizingembedding"><i class="fa fa-check"></i><b>11.3.6</b> Practice Visualizing/Embedding</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="12" data-path="batch-effects.html"><a href="batch-effects.html"><i class="fa fa-check"></i><b>12</b> Batch Effects</a></li>
<li class="chapter" data-level="13" data-path="correcting-batch-effects.html"><a href="correcting-batch-effects.html"><i class="fa fa-check"></i><b>13</b> Correcting Batch Effects</a><ul>
<li class="chapter" data-level="13.1" data-path="correcting-batch-effects.html"><a href="correcting-batch-effects.html#load-settings-and-packages"><i class="fa fa-check"></i><b>13.1</b> Load settings and packages</a></li>
<li class="chapter" data-level="13.2" data-path="correcting-batch-effects.html"><a href="correcting-batch-effects.html#preparing-the-individual-seurat-objects-for-each-pancreas-dataset-without-batch-correction"><i class="fa fa-check"></i><b>13.2</b> Preparing the individual Seurat objects for each pancreas dataset without batch correction</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="cluster-pancreatic-datasets-without-batch-correction.html"><a href="cluster-pancreatic-datasets-without-batch-correction.html"><i class="fa fa-check"></i><b>14</b> Cluster pancreatic datasets without batch correction</a><ul>
<li class="chapter" data-level="14.0.1" data-path="cluster-pancreatic-datasets-without-batch-correction.html"><a href="cluster-pancreatic-datasets-without-batch-correction.html#batch-correction-canonical-correlation-analysis-cca-using-seurat"><i class="fa fa-check"></i><b>14.0.1</b> Batch correction: canonical correlation analysis (CCA) using Seurat</a></li>
<li class="chapter" data-level="14.0.2" data-path="cluster-pancreatic-datasets-without-batch-correction.html"><a href="cluster-pancreatic-datasets-without-batch-correction.html#batch-correction-integrative-non-negative-matrix-factorization-nmf-using-liger"><i class="fa fa-check"></i><b>14.0.2</b> Batch correction: integrative non-negative matrix factorization (NMF) using LIGER</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="additional-exploration-regressing-out-unwanted-covariates.html"><a href="additional-exploration-regressing-out-unwanted-covariates.html"><i class="fa fa-check"></i><b>15</b> Additional exploration: Regressing out unwanted covariates</a></li>
<li class="chapter" data-level="16" data-path="additional-exploration-kbet.html"><a href="additional-exploration-kbet.html"><i class="fa fa-check"></i><b>16</b> Additional exploration: kBET</a></li>
<li class="chapter" data-level="17" data-path="additional-exploration-seurat-3.html"><a href="additional-exploration-seurat-3.html"><i class="fa fa-check"></i><b>17</b> Additional exploration: Seurat 3</a></li>
<li class="chapter" data-level="18" data-path="acknowledgements.html"><a href="acknowledgements.html"><i class="fa fa-check"></i><b>18</b> Acknowledgements</a></li>
<li class="chapter" data-level="19" data-path="functional-analysis.html"><a href="functional-analysis.html"><i class="fa fa-check"></i><b>19</b> Functional Analysis</a><ul>
<li class="chapter" data-level="19.1" data-path="functional-analysis.html"><a href="functional-analysis.html#google-slides-1"><i class="fa fa-check"></i><b>19.1</b> Google Slides</a></li>
<li class="chapter" data-level="19.2" data-path="functional-analysis.html"><a href="functional-analysis.html#gene-sets-and-signatures"><i class="fa fa-check"></i><b>19.2</b> Gene sets and signatures</a><ul>
<li class="chapter" data-level="19.2.1" data-path="functional-analysis.html"><a href="functional-analysis.html#cell-cycle"><i class="fa fa-check"></i><b>19.2.1</b> Cell Cycle</a></li>
</ul></li>
<li class="chapter" data-level="19.3" data-path="functional-analysis.html"><a href="functional-analysis.html#pathway-analysis"><i class="fa fa-check"></i><b>19.3</b> Pathway analysis</a></li>
<li class="chapter" data-level="19.4" data-path="functional-analysis.html"><a href="functional-analysis.html#infercnv-honeybadger"><i class="fa fa-check"></i><b>19.4</b> inferCNV / honeybadger</a><ul>
<li class="chapter" data-level="19.4.1" data-path="functional-analysis.html"><a href="functional-analysis.html#create-the-infercnv-object"><i class="fa fa-check"></i><b>19.4.1</b> Create the InferCNV Object</a></li>
<li class="chapter" data-level="19.4.2" data-path="functional-analysis.html"><a href="functional-analysis.html#filtering-genes"><i class="fa fa-check"></i><b>19.4.2</b> Filtering genes</a></li>
<li class="chapter" data-level="19.4.3" data-path="functional-analysis.html"><a href="functional-analysis.html#normalize-each-cells-counts-for-sequencing-depth"><i class="fa fa-check"></i><b>19.4.3</b> Normalize each cell’s counts for sequencing depth</a></li>
<li class="chapter" data-level="19.4.4" data-path="functional-analysis.html"><a href="functional-analysis.html#perform-anscombe-normalization"><i class="fa fa-check"></i><b>19.4.4</b> Perform Anscombe normalization</a></li>
<li class="chapter" data-level="19.4.5" data-path="functional-analysis.html"><a href="functional-analysis.html#log-transform-the-normalized-counts"><i class="fa fa-check"></i><b>19.4.5</b> Log transform the normalized counts:</a></li>
<li class="chapter" data-level="19.4.6" data-path="functional-analysis.html"><a href="functional-analysis.html#apply-maximum-bounds-to-the-expression-data-to-reduce-outlier-effects"><i class="fa fa-check"></i><b>19.4.6</b> Apply maximum bounds to the expression data to reduce outlier effects</a></li>
<li class="chapter" data-level="19.4.7" data-path="functional-analysis.html"><a href="functional-analysis.html#initial-view-before-infercnv-operations"><i class="fa fa-check"></i><b>19.4.7</b> Initial view, before inferCNV operations:</a></li>
<li class="chapter" data-level="19.4.8" data-path="functional-analysis.html"><a href="functional-analysis.html#perform-smoothing-across-chromosomes"><i class="fa fa-check"></i><b>19.4.8</b> Perform smoothing across chromosomes</a></li>
<li class="chapter" data-level="19.4.9" data-path="functional-analysis.html"><a href="functional-analysis.html#subtract-the-reference-values-from-observations-now-have-logfold-change-values"><i class="fa fa-check"></i><b>19.4.9</b> Subtract the reference values from observations, now have log(fold change) values</a></li>
<li class="chapter" data-level="19.4.10" data-path="functional-analysis.html"><a href="functional-analysis.html#invert-log-values"><i class="fa fa-check"></i><b>19.4.10</b> Invert log values</a></li>
<li class="chapter" data-level="19.4.11" data-path="functional-analysis.html"><a href="functional-analysis.html#removing-noise"><i class="fa fa-check"></i><b>19.4.11</b> Removing noise</a></li>
<li class="chapter" data-level="19.4.12" data-path="functional-analysis.html"><a href="functional-analysis.html#remove-outlier-data-points"><i class="fa fa-check"></i><b>19.4.12</b> Remove outlier data points</a></li>
<li class="chapter" data-level="19.4.13" data-path="functional-analysis.html"><a href="functional-analysis.html#find-de-genes-by-comparing-the-mutant-types-to-normal-types-basic"><i class="fa fa-check"></i><b>19.4.13</b> Find DE genes by comparing the mutant types to normal types, BASIC</a></li>
<li class="chapter" data-level="19.4.14" data-path="functional-analysis.html"><a href="functional-analysis.html#additional-information"><i class="fa fa-check"></i><b>19.4.14</b> Additional Information</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="20" data-path="pseudotime-cell-trajectories.html"><a href="pseudotime-cell-trajectories.html"><i class="fa fa-check"></i><b>20</b> Pseudotime Cell Trajectories</a><ul>
<li class="chapter" data-level="20.1" data-path="pseudotime-cell-trajectories.html"><a href="pseudotime-cell-trajectories.html#google-slides-2"><i class="fa fa-check"></i><b>20.1</b> Google Slides</a></li>
<li class="chapter" data-level="20.2" data-path="pseudotime-cell-trajectories.html"><a href="pseudotime-cell-trajectories.html#comparison-abstract"><i class="fa fa-check"></i><b>20.2</b> Comparison Abstract</a></li>
</ul></li>
<li class="chapter" data-level="21" data-path="functional-pseudotime-analysis.html"><a href="functional-pseudotime-analysis.html"><i class="fa fa-check"></i><b>21</b> Functional Pseudotime Analysis</a></li>
<li class="chapter" data-level="22" data-path="single-cell-multiomic-technologies.html"><a href="single-cell-multiomic-technologies.html"><i class="fa fa-check"></i><b>22</b> Single Cell Multiomic Technologies</a></li>
<li class="chapter" data-level="23" data-path="cite-seq-and-scatac-seq.html"><a href="cite-seq-and-scatac-seq.html"><i class="fa fa-check"></i><b>23</b> CITE-seq and scATAC-seq</a></li>
<li class="chapter" data-level="24" data-path="single-cell-resources.html"><a href="single-cell-resources.html"><i class="fa fa-check"></i><b>24</b> Single Cell Resources</a><ul>
<li class="chapter" data-level="24.1" data-path="single-cell-resources.html"><a href="single-cell-resources.html#comprehensive-list-of-single-cell-resources"><i class="fa fa-check"></i><b>24.1</b> Comprehensive list of single-cell resources</a></li>
<li class="chapter" data-level="24.2" data-path="single-cell-resources.html"><a href="single-cell-resources.html#computational-packages-for-single-cell-analysis"><i class="fa fa-check"></i><b>24.2</b> Computational packages for single-cell analysis</a></li>
<li class="chapter" data-level="24.3" data-path="single-cell-resources.html"><a href="single-cell-resources.html#elife-commentary-on-the-human-cell-atlas"><i class="fa fa-check"></i><b>24.3</b> eLife Commentary on the Human Cell Atlas</a></li>
<li class="chapter" data-level="24.4" data-path="single-cell-resources.html"><a href="single-cell-resources.html#online-courses"><i class="fa fa-check"></i><b>24.4</b> Online courses</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="feature-selection-and-cluster-analysis" class="section level1">
<h1><span class="header-section-number">11</span> Feature Selection and Cluster Analysis</h1>
<div id="abstract" class="section level2">
<h2><span class="header-section-number">11.1</span> Abstract</h2>
<p>Many methods have been used to determine differential gene expression from single-cell RNA (scRNA)-seq data. We evaluated 36 approaches using experimental and synthetic data and found considerable differences in the number and characteristics of the genes that are called differentially expressed. Prefiltering of lowly expressed genes has important effects, particularly for some of the methods developed for bulk RNA-seq data analysis. However, we found that bulk RNA-seq analysis methods do not generally perform worse than those developed specifically for scRNA-seq. We also present conquer, a repository of consistently processed, analysis-ready public scRNA-seq data sets that is aimed at simplifying method evaluation and reanalysis of published results. Each data set provides abundance estimates for both genes and transcripts, as well as quality control and exploratory analysis reports. <span class="citation">(Soneson and Robinson <a href="#ref-soneson2018bias">2018</a>)</span></p>
<p>Cells are the basic building blocks of organisms and each cell is unique. Single-cell RNA sequencing has emerged as an indispensable tool to dissect the cellular heterogeneity and decompose tissues into cell types and/or cell states, which offers enormous potential for de novo discovery. Single-cell transcriptomic atlases provide unprecedented resolution to reveal complex cellular events and deepen our understanding of biological systems. In this review, we summarize and compare single-cell RNA sequencing technologies, that were developed since 2009, to facilitate a well-informed choice of method. The applications of these methods in different biological contexts are also discussed. We anticipate an ever-increasing role of single-cell RNA sequencing in biology with further improvement in providing spatial information and coupling to other cellular modalities. In the future, such biological findings will greatly benefit medical research. <span class="citation">(Hedlund and Deng <a href="#ref-hedlund2018single">2018</a>)</span></p>
</div>
<div id="seurat-tutorial" class="section level2">
<h2><span class="header-section-number">11.2</span> Seurat Tutorial</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(Seurat)
<span class="kw">library</span>(dplyr)
<span class="kw">library</span>(ggplot2)
<span class="kw">library</span>(CountClust)

pbmc &lt;-<span class="st"> </span>pbmc_small
pbmc<span class="op">@</span>meta.data<span class="op">$</span>barcode &lt;-<span class="st"> </span><span class="kw">row.names</span>(pbmc<span class="op">@</span>meta.data)
<span class="co"># pbmc@ident &lt;- factor()</span></code></pre></div>
<div id="preprocessing-steps" class="section level3">
<h3><span class="header-section-number">11.2.1</span> Preprocessing Steps</h3>
<p>This was all covered in the last Lab!</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># The number of genes and UMIs (nGene and nUMI) are automatically calculated</span>
<span class="co"># for every object by Seurat.  For non-UMI data, nUMI represents the sum of</span>
<span class="co"># the non-normalized values within a cell We calculate the percentage of</span>
<span class="co"># mitochondrial genes here and store it in percent.mito using AddMetaData.</span>
<span class="co"># We use object@raw.data since this represents non-transformed and</span>
<span class="co"># non-log-normalized counts The % of UMI mapping to MT-genes is a common</span>
<span class="co"># scRNA-seq QC metric.</span>
mito.genes &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="dt">pattern =</span> <span class="st">&quot;^MT-&quot;</span>, <span class="dt">x =</span> <span class="kw">rownames</span>(<span class="dt">x =</span> pbmc<span class="op">@</span>data), <span class="dt">value =</span> <span class="ot">TRUE</span>)
percent.mito &lt;-<span class="st"> </span>Matrix<span class="op">::</span><span class="kw">colSums</span>(pbmc<span class="op">@</span>raw.data[mito.genes, ])<span class="op">/</span>Matrix<span class="op">::</span><span class="kw">colSums</span>(pbmc<span class="op">@</span>raw.data)

<span class="co"># AddMetaData adds columns to object@meta.data, and is a great place to</span>
<span class="co"># stash QC stats</span>
pbmc &lt;-<span class="st"> </span><span class="kw">AddMetaData</span>(<span class="dt">object =</span> pbmc, <span class="dt">metadata =</span> percent.mito, <span class="dt">col.name =</span> <span class="st">&quot;percent.mito&quot;</span>)
##VlnPlot(object = pbmc, features.plot = c(&quot;nGene&quot;, &quot;nUMI&quot;), &quot;percent.mito&quot;), nCol = 3)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># GenePlot is typically used to visualize gene-gene relationships, but can</span>
<span class="co"># be used for anything calculated by the object, i.e. columns in</span>
<span class="co"># object@meta.data, PC scores etc.  Since there is a rare subset of cells</span>
<span class="co"># with an outlier level of high mitochondrial percentage and also low UMI</span>
<span class="co"># content, we filter these as well</span>
<span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))
<span class="kw">GenePlot</span>(<span class="dt">object =</span> pbmc, <span class="dt">gene1 =</span> <span class="st">&quot;nUMI&quot;</span>, <span class="dt">gene2 =</span> <span class="st">&quot;percent.mito&quot;</span>)
<span class="kw">GenePlot</span>(<span class="dt">object =</span> pbmc, <span class="dt">gene1 =</span> <span class="st">&quot;nUMI&quot;</span>, <span class="dt">gene2 =</span> <span class="st">&quot;nGene&quot;</span>)</code></pre></div>
<p><strong>Already A Subset of the Data</strong></p>
<p>This is not run becuase we are already working with a subset. This is here for reference.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># We filter out cells that have unique gene counts over 2,500 or less than</span>
<span class="co"># 200 Note that low.thresholds and high.thresholds are used to define a</span>
<span class="co"># &#39;gate&#39;.  -Inf and Inf should be used if you don&#39;t want a lower or upper</span>
<span class="co"># threshold.</span>

pbmc &lt;-<span class="st"> </span><span class="kw">FilterCells</span>(<span class="dt">object =</span> pbmc, 
                    <span class="dt">subset.names =</span> <span class="kw">c</span>(<span class="st">&quot;nGene&quot;</span>, <span class="st">&quot;percent.mito&quot;</span>), 
                    <span class="dt">low.thresholds =</span> <span class="kw">c</span>(<span class="dv">200</span>, <span class="op">-</span><span class="ot">Inf</span>), 
                    <span class="dt">high.thresholds =</span> <span class="kw">c</span>(<span class="dv">2500</span>, <span class="fl">0.1</span>))</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pbmc &lt;-<span class="st"> </span><span class="kw">NormalizeData</span>(<span class="dt">object =</span> pbmc, <span class="dt">normalization.method =</span> <span class="st">&quot;LogNormalize&quot;</span>, <span class="dt">scale.factor =</span> <span class="dv">10000</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pbmc &lt;-<span class="st"> </span><span class="kw">FindVariableGenes</span>(<span class="dt">object =</span> pbmc, 
                          <span class="dt">mean.function =</span> ExpMean, 
                          <span class="dt">dispersion.function =</span> LogVMR, 
                          <span class="dt">do.plot =</span> <span class="ot">FALSE</span>)</code></pre></div>
</div>
<div id="start-of-identifying-cell-types" class="section level3">
<h3><span class="header-section-number">11.2.2</span> Start of Identifying Cell Types</h3>
<div id="scaling" class="section level4">
<h4><span class="header-section-number">11.2.2.1</span> Scaling</h4>
<p>This part is where you mean center the data, substract the mean. You also divide by the standard deviation to make everything to a ‘standard normal’, where the mean is zero and the standard deviation is 1.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pbmc &lt;-<span class="st"> </span><span class="kw">ScaleData</span>(<span class="dt">object =</span> pbmc, <span class="dt">vars.to.regress =</span> <span class="kw">c</span>(<span class="st">&quot;batch&quot;</span>, <span class="st">&quot;percent.mito&quot;</span>))</code></pre></div>
<p>Try Regressing Other Variables</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## batch_ids &lt;- read.csv(&#39;file_with_batch_ids.csv&#39;)

## randomly making a batch id data.frame
batch_ids &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">barcode =</span> <span class="kw">rownames</span>(pbmc<span class="op">@</span>meta.data), 
                        <span class="dt">batch_id =</span> <span class="kw">sample</span>(<span class="dv">0</span><span class="op">:</span><span class="dv">2</span>, <span class="dv">80</span>, <span class="dt">replace =</span> <span class="ot">TRUE</span>),
                        <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)


<span class="kw">row.names</span>(batch_ids) &lt;-<span class="st"> </span><span class="kw">row.names</span>(pbmc<span class="op">@</span>meta.data)

pbmc &lt;-<span class="st"> </span><span class="kw">AddMetaData</span>(<span class="dt">object =</span> pbmc, <span class="dt">metadata =</span> batch_ids, <span class="dt">col.name =</span> <span class="st">&quot;batch_id&quot;</span>)

pbmc &lt;-<span class="st"> </span><span class="kw">ScaleData</span>(<span class="dt">object =</span> pbmc, <span class="dt">vars.to.regress =</span> <span class="st">&#39;batch_id&#39;</span>)</code></pre></div>
</div>
<div id="running-pca" class="section level4">
<h4><span class="header-section-number">11.2.2.2</span> Running PCA</h4>
<p>This will run pca on the</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pbmc &lt;-<span class="st"> </span><span class="kw">RunPCA</span>(<span class="dt">object =</span> pbmc, 
               <span class="dt">pc.genes =</span> pbmc<span class="op">@</span>var.genes, 
               <span class="dt">do.print =</span> <span class="ot">TRUE</span>, 
               <span class="dt">pcs.print =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, 
               <span class="dt">genes.print =</span> <span class="dv">5</span>)</code></pre></div>
</div>
<div id="running-ica" class="section level4">
<h4><span class="header-section-number">11.2.2.3</span> Running ICA</h4>
<p>Try running Independent Component Analysis. If you need help with the inputs try using the ?RunICA menu.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pbmc &lt;-<span class="st"> </span><span class="kw">RunICA</span>(pbmc, <span class="dt">ics.compute=</span><span class="dv">5</span>)</code></pre></div>
</div>
<div id="visualizing-pca-in-different-ways" class="section level4">
<h4><span class="header-section-number">11.2.2.4</span> Visualizing PCA in Different Ways</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">VizPCA</span>(<span class="dt">object =</span> pbmc, <span class="dt">pcs.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>)</code></pre></div>
</div>
<div id="visualizing-ica-in-different-ways" class="section level4">
<h4><span class="header-section-number">11.2.2.5</span> Visualizing ICA in Different Ways</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">VizICA</span>(<span class="dt">object =</span> pbmc, <span class="dt">ics.use=</span><span class="dv">1</span><span class="op">:</span><span class="dv">3</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">PCAPlot</span>(<span class="dt">object =</span> pbmc, <span class="dt">dim.1 =</span> <span class="dv">1</span>, <span class="dt">dim.2 =</span> <span class="dv">2</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># ProjectPCA scores each gene in the dataset (including genes not included</span>
<span class="co"># in the PCA) based on their correlation with the calculated components.</span>
<span class="co"># Though we don&#39;t use this further here, it can be used to identify markers</span>
<span class="co"># that are strongly correlated with cellular heterogeneity, but may not have</span>
<span class="co"># passed through variable gene selection.  The results of the projected PCA</span>
<span class="co"># can be explored by setting use.full=T in the functions above</span>
pbmc &lt;-<span class="st"> </span><span class="kw">ProjectPCA</span>(<span class="dt">object =</span> pbmc, <span class="dt">do.print =</span> <span class="ot">FALSE</span>)</code></pre></div>
</div>
<div id="genes-by-pcs" class="section level4">
<h4><span class="header-section-number">11.2.2.6</span> Genes by PCs</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">PCHeatmap</span>(<span class="dt">object =</span> pbmc, 
          <span class="dt">pc.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, 
          <span class="dt">cells.use =</span> <span class="dv">50</span>, 
          <span class="dt">do.balanced =</span> <span class="ot">TRUE</span>, 
          <span class="dt">label.columns =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>Check other PCs to plot</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">PCHeatmap</span>()</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">PCElbowPlot</span>(<span class="dt">object =</span> pbmc, <span class="dt">num.pc =</span> <span class="dv">10</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># save.SNN = T saves the SNN so that the clustering algorithm can be rerun</span>
<span class="co"># using the same graph but with a different resolution value (see docs for</span>
<span class="co"># full details)</span>
<span class="kw">set.seed</span>(<span class="dv">2019</span>)
pbmc &lt;-<span class="st"> </span><span class="kw">FindClusters</span>(<span class="dt">object =</span> pbmc, 
                     <span class="dt">reduction.type =</span> <span class="st">&quot;pca&quot;</span>, 
                     <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, 
                     <span class="dt">resolution =</span> <span class="dv">1</span>, 
                     <span class="dt">print.output =</span> <span class="dv">0</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="kw">runif</span>(<span class="dv">100</span>))
pbmc &lt;-<span class="kw">RunTSNE</span>(pbmc, <span class="dt">reduction.use =</span> <span class="st">&quot;pca&quot;</span>, <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dt">perplexity=</span><span class="dv">10</span>)

<span class="co"># note that you can set do.label=T to help label individual clusters</span>
<span class="kw">TSNEPlot</span>(<span class="dt">object =</span> pbmc)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># find all markers of cluster 1</span>
cluster1.markers &lt;-<span class="st"> </span><span class="kw">FindMarkers</span>(<span class="dt">object =</span> pbmc, <span class="dt">ident.1 =</span> <span class="dv">1</span>, <span class="dt">min.pct =</span> <span class="fl">0.25</span>)
<span class="kw">print</span>(<span class="dt">x =</span> <span class="kw">head</span>(<span class="dt">x =</span> cluster1.markers, <span class="dt">n =</span> <span class="dv">5</span>))</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># find all markers distinguishing cluster 2 from clusters 0 and 1</span>
cluster2.markers &lt;-<span class="st"> </span><span class="kw">FindMarkers</span>(<span class="dt">object =</span> pbmc, <span class="dt">ident.1 =</span> <span class="dv">2</span>, <span class="dt">ident.2 =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">min.pct =</span> <span class="fl">0.25</span>)
<span class="kw">print</span>(<span class="dt">x =</span> <span class="kw">head</span>(<span class="dt">x =</span> cluster5.markers, <span class="dt">n =</span> <span class="dv">5</span>))</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># find markers for every cluster compared to all remaining cells, report</span>
<span class="co"># only the positive ones</span>
pbmc.markers &lt;-<span class="st"> </span><span class="kw">FindAllMarkers</span>(<span class="dt">object =</span> pbmc, <span class="dt">only.pos =</span> <span class="ot">TRUE</span>, <span class="dt">min.pct =</span> <span class="fl">0.25</span>, <span class="dt">thresh.use =</span> <span class="fl">0.25</span>)
pbmc.markers <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(cluster) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">top_n</span>(<span class="dv">2</span>, avg_logFC)</code></pre></div>
</div>
<div id="find-marker-genes" class="section level4">
<h4><span class="header-section-number">11.2.2.7</span> Find Marker Genes</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cluster1.markers &lt;-<span class="st"> </span><span class="kw">FindMarkers</span>(<span class="dt">object =</span> pbmc, 
                                <span class="dt">ident.1 =</span> <span class="dv">0</span>, 
                                <span class="dt">thresh.use =</span> <span class="fl">0.25</span>, 
                                <span class="dt">test.use =</span> <span class="st">&quot;roc&quot;</span>, 
                                <span class="dt">only.pos =</span> <span class="ot">TRUE</span>)


<span class="kw">VlnPlot</span>(<span class="dt">object =</span> pbmc, <span class="dt">features.plot =</span> <span class="kw">c</span>(<span class="st">&quot;MS4A1&quot;</span>, <span class="st">&quot;CD79A&quot;</span>))</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># you can plot raw UMI counts as well</span>
<span class="kw">VlnPlot</span>(<span class="dt">object =</span> pbmc, 
        <span class="dt">features.plot =</span> <span class="kw">c</span>(<span class="st">&quot;NKG7&quot;</span>, <span class="st">&quot;PF4&quot;</span>), 
        <span class="dt">use.raw =</span> <span class="ot">TRUE</span>, 
        <span class="dt">y.log =</span> <span class="ot">TRUE</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">FeaturePlot</span>(<span class="dt">object =</span> pbmc, 
            <span class="dt">features.plot =</span> <span class="kw">c</span>(<span class="st">&quot;MS4A1&quot;</span>, <span class="st">&quot;GNLY&quot;</span>, <span class="st">&quot;CD3E&quot;</span>, <span class="st">&quot;CD14&quot;</span>, <span class="st">&quot;FCER1A&quot;</span>, <span class="st">&quot;FCGR3A&quot;</span>, <span class="st">&quot;LYZ&quot;</span>, <span class="st">&quot;PPBP&quot;</span>, <span class="st">&quot;CD8A&quot;</span>),
            <span class="dt">cols.use =</span> <span class="kw">c</span>(<span class="st">&quot;grey&quot;</span>, <span class="st">&quot;blue&quot;</span>), 
            <span class="dt">sreduction.use =</span> <span class="st">&quot;tsne&quot;</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">top10 &lt;-<span class="st"> </span>pbmc.markers <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(cluster) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">top_n</span>(<span class="dv">10</span>, avg_logFC)
<span class="co"># setting slim.col.label to TRUE will print just the cluster IDS instead of</span>
<span class="co"># every cell name</span>
<span class="kw">DoHeatmap</span>(<span class="dt">object =</span> pbmc, <span class="dt">genes.use =</span> top10<span class="op">$</span>gene, <span class="dt">slim.col.label =</span> <span class="ot">TRUE</span>, <span class="dt">remove.key =</span> <span class="ot">TRUE</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">current.cluster.ids &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">6</span>, <span class="dv">7</span>)
new.cluster.ids &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;CD4 T cells&quot;</span>, <span class="st">&quot;CD14+ Monocytes&quot;</span>, <span class="st">&quot;B cells&quot;</span>, <span class="st">&quot;CD8 T cells&quot;</span>, <span class="st">&quot;FCGR3A+ Monocytes&quot;</span>, <span class="st">&quot;NK cells&quot;</span>, <span class="st">&quot;Dendritic cells&quot;</span>, <span class="st">&quot;Megakaryocytes&quot;</span>)
pbmc<span class="op">@</span>ident &lt;-<span class="st"> </span>plyr<span class="op">::</span><span class="kw">mapvalues</span>(<span class="dt">x =</span> pbmc<span class="op">@</span>ident, <span class="dt">from =</span> current.cluster.ids, <span class="dt">to =</span> new.cluster.ids)
<span class="kw">TSNEPlot</span>(<span class="dt">object =</span> pbmc, <span class="dt">do.label =</span> <span class="ot">TRUE</span>, <span class="dt">pt.size =</span> <span class="fl">0.5</span>)</code></pre></div>
</div>
<div id="further-subdivisions-within-cell-types" class="section level4">
<h4><span class="header-section-number">11.2.2.8</span> Further subdivisions within cell types</h4>
<p>If you perturb some of our parameter choices above (for example, setting resolution=0.8 or changing the number of PCs), you might see the CD4 T cells subdivide into two groups. You can explore this subdivision to find markers separating the two T cell subsets. However, before reclustering (which will overwrite <a href="mailto:object@ident">object@ident</a>), we can stash our renamed identities to be easily recovered later.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># First lets stash our identities for later</span>
pbmc &lt;-<span class="st"> </span><span class="kw">StashIdent</span>(<span class="dt">object =</span> pbmc, <span class="dt">save.name =</span> <span class="st">&quot;ClusterNames_0.6&quot;</span>)
<span class="co"># Note that if you set save.snn=T above, you don&#39;t need to recalculate the</span>
<span class="co"># SNN, and can simply put: pbmc &lt;- FindClusters(pbmc,resolution = 0.8)</span>
pbmc &lt;-<span class="st"> </span><span class="kw">FindClusters</span>(<span class="dt">object =</span> pbmc, 
                     <span class="dt">reduction.type =</span> <span class="st">&quot;pca&quot;</span>, 
                     <span class="dt">dims.use =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, 
                     <span class="dt">resolution =</span> <span class="fl">0.8</span>, 
                     <span class="dt">print.output =</span> <span class="ot">FALSE</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Warning in BuildSNN(object = object, genes.use = genes.use, reduction.type
## = reduction.type, : Build parameters exactly match those of already
## computed and stored SNN. To force recalculation, set force.recalc to TRUE.
<span class="co"># Demonstration of how to plot two tSNE plots side by side, and how to color</span>
<span class="co"># points based on different criteria</span>
plot1 &lt;-<span class="st"> </span><span class="kw">TSNEPlot</span>(<span class="dt">object =</span> pbmc, 
                  <span class="dt">do.return =</span> <span class="ot">TRUE</span>, 
                  <span class="dt">no.legend =</span> <span class="ot">TRUE</span>, 
                  <span class="dt">do.label =</span> <span class="ot">TRUE</span>)

plot2 &lt;-<span class="st"> </span><span class="kw">TSNEPlot</span>(<span class="dt">object =</span> pbmc, 
                  <span class="dt">do.return =</span> <span class="ot">TRUE</span>, 
                  <span class="dt">group.by =</span> <span class="st">&quot;ClusterNames_0.6&quot;</span>, 
                  <span class="dt">no.legend =</span> <span class="ot">TRUE</span>, 
                  <span class="dt">do.label =</span> <span class="ot">TRUE</span>)

<span class="kw">plot_grid</span>(plot1, plot2)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Find discriminating markers</span>
tcell.markers &lt;-<span class="st"> </span><span class="kw">FindMarkers</span>(<span class="dt">object =</span> pbmc, <span class="dt">ident.1 =</span> <span class="dv">0</span>, <span class="dt">ident.2 =</span> <span class="dv">1</span>)

<span class="co"># Most of the markers tend to be expressed in C1 (i.e. S100A4). However, we</span>
<span class="co"># can see that CCR7 is upregulated in C0, strongly indicating that we can</span>
<span class="co"># differentiate memory from naive CD4 cells.  cols.use demarcates the color</span>
<span class="co"># palette from low to high expression</span>
<span class="kw">FeaturePlot</span>(<span class="dt">object =</span> pbmc, <span class="dt">features.plot =</span> <span class="kw">c</span>(<span class="st">&quot;S100A4&quot;</span>, <span class="st">&quot;CCR7&quot;</span>), <span class="dt">cols.use =</span> <span class="kw">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;blue&quot;</span>))</code></pre></div>
</div>
</div>
</div>
<div id="feature-selection" class="section level2">
<h2><span class="header-section-number">11.3</span> Feature Selection</h2>
<div id="differential-expression-analysis" class="section level3">
<h3><span class="header-section-number">11.3.1</span> Differential Expression Analysis</h3>
<div id="differential-expression-tests" class="section level4">
<h4><span class="header-section-number">11.3.1.1</span> Differential Expression Tests</h4>
<p>One of the most commonly performed tasks for RNA-seq data is differential gene expression (DE) analysis. Although well- established tools exist for such analysis in bulk RNA-seq data6–8, methods for scRNA-seq data are just emerging. Given the special characteristics of scRNA-seq data, including generally low library sizes, high noise levels and a large fraction of so-called ‘dropout’ events, it is unclear whether DE methods that have been devel- oped for bulk RNA-seq are suitable also for scRNA-seq</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Differential expression using DESeq2
<span class="kw">DESeq2DETest</span>(<span class="dt">object =</span> pbmc, 
             cells.<span class="dv">1</span>, 
             cells.<span class="dv">2</span>, 
             <span class="dt">genes.use =</span> <span class="ot">NULL</span>, 
             <span class="dt">assay.type =</span> <span class="st">&quot;RNA&quot;</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Likelihood ratio test for zero-inflated data
<span class="kw">DiffExpTest</span>(<span class="dt">object =</span> pbmc, 
            cells.<span class="dv">1</span>, 
            cells.<span class="dv">2</span>, 
            <span class="dt">assay.type =</span> <span class="st">&quot;RNA&quot;</span>, 
            <span class="dt">genes.use =</span> <span class="ot">NULL</span>,
            <span class="dt">print.bar =</span> <span class="ot">TRUE</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## t-test
<span class="kw">DiffTTest</span>(<span class="dt">object =</span> pbmc, 
          cells.<span class="dv">1</span>, 
          cells.<span class="dv">2</span>, 
          <span class="dt">genes.use =</span> <span class="ot">NULL</span>, 
          <span class="dt">print.bar =</span> <span class="ot">TRUE</span>,
          <span class="dt">assay.type =</span> <span class="st">&quot;RNA&quot;</span>)</code></pre></div>
</div>
</div>
<div id="dimensionality-reduction" class="section level3">
<h3><span class="header-section-number">11.3.2</span> Dimensionality Reduction</h3>
<div id="principal-components-analysis-pca" class="section level4">
<h4><span class="header-section-number">11.3.2.1</span> Principal Components Analysis (PCA)</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">RunPCA</span>()</code></pre></div>
</div>
</div>
<div id="independent-components-analysis-ica" class="section level3">
<h3><span class="header-section-number">11.3.3</span> Independent Components Analysis (ICA)</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">RunICA</span>()</code></pre></div>
</div>
<div id="clustering" class="section level3">
<h3><span class="header-section-number">11.3.4</span> Clustering</h3>
<div id="kmeans" class="section level4">
<h4><span class="header-section-number">11.3.4.1</span> Kmeans</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">DoKMeans</span>(object, <span class="dt">genes.use =</span> <span class="ot">NULL</span>, <span class="dt">k.genes =</span> <span class="ot">NULL</span>, <span class="dt">k.cells =</span> <span class="dv">0</span>,
<span class="dt">k.seed =</span> <span class="dv">1</span>, <span class="dt">do.plot =</span> <span class="ot">FALSE</span>, <span class="dt">data.cut =</span> <span class="fl">2.5</span>,
<span class="dt">k.cols =</span> <span class="kw">PurpleAndYellow</span>(), <span class="dt">set.ident =</span> <span class="ot">TRUE</span>, <span class="dt">do.constrained =</span> <span class="ot">FALSE</span>,
<span class="dt">assay.type =</span> <span class="st">&quot;RNA&quot;</span>, ...)</code></pre></div>
</div>
<div id="louvain" class="section level4">
<h4><span class="header-section-number">11.3.4.2</span> Louvain</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">##  Neighborhood graph
<span class="kw">BuildSNN</span>(object, <span class="dt">genes.use =</span> <span class="ot">NULL</span>, <span class="dt">reduction.type =</span> <span class="st">&quot;pca&quot;</span>,
<span class="dt">dims.use =</span> <span class="ot">NULL</span>, <span class="dt">k.param =</span> <span class="dv">10</span>, <span class="dt">plot.SNN =</span> <span class="ot">FALSE</span>, <span class="dt">prune.SNN =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">15</span>,
<span class="dt">print.output =</span> <span class="ot">TRUE</span>, <span class="dt">distance.matrix =</span> <span class="ot">NULL</span>, <span class="dt">force.recalc =</span> <span class="ot">FALSE</span>,
<span class="dt">filename =</span> <span class="ot">NULL</span>, <span class="dt">save.SNN =</span> <span class="ot">TRUE</span>, <span class="dt">nn.eps =</span> <span class="dv">0</span>)

<span class="kw">FindClusters</span>(object, <span class="dt">genes.use =</span> <span class="ot">NULL</span>, <span class="dt">reduction.type =</span> <span class="st">&quot;pca&quot;</span>,
<span class="dt">dims.use =</span> <span class="ot">NULL</span>, <span class="dt">k.param =</span> <span class="dv">30</span>, <span class="dt">plot.SNN =</span> <span class="ot">FALSE</span>, <span class="dt">prune.SNN =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">15</span>,
<span class="dt">print.output =</span> <span class="ot">TRUE</span>, <span class="dt">distance.matrix =</span> <span class="ot">NULL</span>, <span class="dt">save.SNN =</span> <span class="ot">FALSE</span>,
<span class="dt">reuse.SNN =</span> <span class="ot">FALSE</span>, <span class="dt">force.recalc =</span> <span class="ot">FALSE</span>, <span class="dt">nn.eps =</span> <span class="dv">0</span>,
<span class="dt">modularity.fxn =</span> <span class="dv">1</span>, <span class="dt">resolution =</span> <span class="fl">0.8</span>, <span class="dt">algorithm =</span> <span class="dv">1</span>, <span class="dt">n.start =</span> <span class="dv">100</span>,
<span class="dt">n.iter =</span> <span class="dv">10</span>, <span class="dt">random.seed =</span> <span class="dv">0</span>, <span class="dt">temp.file.location =</span> <span class="ot">NULL</span>,
<span class="dt">edge.file.name =</span> <span class="ot">NULL</span>)</code></pre></div>
</div>
<div id="probabilistic-lda" class="section level4">
<h4><span class="header-section-number">11.3.4.3</span> Probabilistic (LDA)</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">2019</span>)
## Set Cluster Identity to the clustering using a resolution of 1 (res.1) 
<span class="kw">SetIdent</span>(pbmc_small, <span class="dt">ident.use =</span> <span class="st">&#39;res.1&#39;</span>)

pbmc_counts &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(pbmc_small<span class="op">@</span>data)
pbmc_meta &lt;-<span class="st"> </span>pbmc_small<span class="op">@</span>meta.data
gene_names &lt;-<span class="st"> </span><span class="kw">rownames</span>(pbmc_counts)


pbmc_FitGoM &lt;-<span class="st"> </span><span class="kw">FitGoM</span>(<span class="kw">t</span>(pbmc_counts), <span class="dt">K=</span><span class="dv">4</span>)


omega &lt;-<span class="st"> </span>pbmc_FitGoM<span class="op">$</span>fit<span class="op">$</span>omega

annotation &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">sample_id =</span> <span class="kw">rownames</span>(omega),
                         <span class="dt">tissue_label =</span> <span class="kw">paste0</span>(<span class="st">&quot;topic&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>))

<span class="kw">colnames</span>(omega) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;topic&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>)
<span class="kw">rownames</span>(omega) &lt;-<span class="st"> </span>annotation<span class="op">$</span>sample_id;

<span class="kw">StructureGGplot</span>(<span class="dt">omega =</span> omega,
                <span class="dt">annotation =</span> annotation,
                <span class="dt">palette =</span> RColorBrewer<span class="op">::</span><span class="kw">brewer.pal</span>(<span class="dv">4</span>, <span class="st">&quot;Dark2&quot;</span>),
                <span class="dt">yaxis_label =</span> <span class="st">&quot;Cells&quot;</span>,
                <span class="dt">order_sample =</span> <span class="ot">TRUE</span>,
                <span class="dt">axis_tick =</span> <span class="kw">list</span>(<span class="dt">axis_ticks_length =</span> .<span class="dv">1</span>,
                                 <span class="dt">axis_ticks_lwd_y =</span> .<span class="dv">1</span>,
                                 <span class="dt">axis_ticks_lwd_x =</span> .<span class="dv">1</span>,
                                 <span class="dt">axis_label_size =</span> <span class="dv">7</span>,
                                 <span class="dt">axis_label_face =</span> <span class="st">&quot;bold&quot;</span>))


## Add Topic Scores to Meta Data Part of the Seurat Object
pbmc_small<span class="op">@</span>meta.data &lt;-<span class="st"> </span><span class="kw">cbind</span>(pbmc_meta, omega)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pbmc_small<span class="op">@</span>meta.data <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">group_by</span>(res.<span class="dv">1</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">summarise</span>(<span class="dt">topic1 =</span> <span class="kw">mean</span>(topic1),
            <span class="dt">topic2 =</span> <span class="kw">mean</span>(topic2),
            <span class="dt">topic3 =</span> <span class="kw">mean</span>(topic3),
            <span class="dt">topic4 =</span> <span class="kw">mean</span>(topic4))</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## ggplot object, you can add layers
p1 &lt;-<span class="st"> </span><span class="kw">TSNEPlot</span>(pbmc_small, <span class="dt">do.return =</span> <span class="ot">TRUE</span>) <span class="op">+</span><span class="st"> </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">&quot;Resolution 1&quot;</span>) ## return ggplot object
p1</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p2 &lt;-<span class="st"> </span><span class="kw">FeaturePlot</span>(<span class="dt">object =</span> pbmc_small, 
                  <span class="dt">features.plot =</span> <span class="kw">c</span>(<span class="st">&quot;topic1&quot;</span>, <span class="st">&quot;topic2&quot;</span>, <span class="st">&quot;topic3&quot;</span>, <span class="st">&quot;topic4&quot;</span>), 
                  <span class="dt">cols.use =</span> <span class="kw">c</span>(<span class="st">&quot;grey&quot;</span>, <span class="st">&quot;blue&quot;</span>), 
                  <span class="dt">reduction.use =</span> <span class="st">&quot;tsne&quot;</span>, <span class="dt">do.return =</span> <span class="ot">TRUE</span>) ## return ggplot object

p2</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot_grid</span>(p1, p2)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(p2)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot_grid</span>(p1, p2[[<span class="dv">1</span>]], p2[[<span class="dv">2</span>]], p2[[<span class="dv">3</span>]], p2[[<span class="dv">4</span>]])</code></pre></div>
</div>
<div id="extract-top-feature" class="section level4">
<h4><span class="header-section-number">11.3.4.4</span> Extract Top Feature</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">theta_mat &lt;-<span class="st"> </span>pbmc_FitGoM<span class="op">$</span>fit<span class="op">$</span>theta

top_features &lt;-<span class="st"> </span><span class="kw">ExtractTopFeatures</span>(theta_mat, 
                                   <span class="dt">top_features=</span><span class="dv">100</span>,
                                   <span class="dt">method=</span><span class="st">&quot;poisson&quot;</span>, 
                                   <span class="dt">options=</span><span class="st">&quot;min&quot;</span>)

gene_list &lt;-<span class="st"> </span><span class="kw">do.call</span>(rbind, 
                     <span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">dim</span>(top_features<span class="op">$</span>indices)[<span class="dv">1</span>],
                            <span class="cf">function</span>(x) gene_names[top_features<span class="op">$</span>indices[x,]]))</code></pre></div>
<p>We tabulate the top <span class="math inline">\(5\)</span> genes for these <span class="math inline">\(6\)</span> clusters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tmp &lt;-<span class="st"> </span><span class="kw">do.call</span>(rbind, <span class="kw">lapply</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="cf">function</span>(i) <span class="kw">toString</span>(gene_list[,i])))
<span class="kw">rownames</span>(tmp) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;Cluster&quot;</span>, <span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>))
tmp <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">kable</span>(<span class="st">&quot;html&quot;</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">kable_styling</span>()</code></pre></div>
</div>
</div>
<div id="check-clusters" class="section level3">
<h3><span class="header-section-number">11.3.5</span> Check Clusters</h3>
<p>Use Classifier to predict cell cluster. See how it predicts using hold out data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">test.pbmc &lt;-<span class="st"> </span><span class="kw">SubsetData</span>(<span class="dt">object =</span> pbmc_small, <span class="dt">cells.use =</span> pbmc_small<span class="op">@</span>cell.names[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>])
train.pbmc &lt;-<span class="st"> </span><span class="kw">SubsetData</span>(<span class="dt">object =</span> pbmc_small, <span class="dt">cells.use =</span> pbmc_small<span class="op">@</span>cell.names[<span class="dv">11</span><span class="op">:</span><span class="dv">80</span>])
predicted.classes &lt;-<span class="st"> </span><span class="kw">ClassifyCells</span>(
  <span class="dt">object =</span> train.pbmc,
  <span class="dt">training.classes =</span> train.pbmc<span class="op">@</span>ident,
  <span class="dt">new.data =</span> test.pbmc<span class="op">@</span>data
)


<span class="kw">table</span>(predicted.classes, test.pbmc<span class="op">@</span>ident)</code></pre></div>
<p>Visually check by comparing centroids of clusters in gene space and embedding space.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pmbc_small &lt;-<span class="st"> </span><span class="kw">GetCentroids</span>(pbmc_small, <span class="dt">cells.use=</span>pbmc_small<span class="op">@</span>cell.names)</code></pre></div>
</div>
<div id="practice-visualizingembedding" class="section level3">
<h3><span class="header-section-number">11.3.6</span> Practice Visualizing/Embedding</h3>
<div id="tsne" class="section level4">
<h4><span class="header-section-number">11.3.6.1</span> tSNE</h4>
<p>Change the parameter settings for tSNE</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">RunTSNE</span>()</code></pre></div>
</div>
<div id="umap" class="section level4">
<h4><span class="header-section-number">11.3.6.2</span> UMAP</h4>
<p>Change the parameter settings for UMAP</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">RunUMAP</span>()</code></pre></div>

</div>
</div>
</div>
</div>
<h3>References</h3>
<div id="refs" class="references">
<div id="ref-soneson2018bias">
<p>Soneson, Charlotte, and Mark D Robinson. 2018. “Bias, Robustness and Scalability in Single-Cell Differential Expression Analysis.” <em>Nature Methods</em> 15 (4). Nature Publishing Group: 255.</p>
</div>
<div id="ref-hedlund2018single">
<p>Hedlund, Eva, and Qiaolin Deng. 2018. “Single-Cell Rna Sequencing: Technical Advancements and Biological Applications.” <em>Molecular Aspects of Medicine</em> 59. Elsevier: 36–46.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="identifying-cell-populations.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="batch-effects.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["2019_scWorkshop.pdf", "2019_scWorkshop.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
