<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>14 Functional Analysis | ANALYSIS OF SINGLE CELL RNA-SEQ DATA</title>
  <meta name="description" content="This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="14 Functional Analysis | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://broadinstitute.github.io/2019_scWorkshop/" />
  <meta property="og:image" content="https://broadinstitute.github.io/2019_scWorkshop/images/cover.png" />
  <meta property="og:description" content="This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook." />
  <meta name="github-repo" content="broadinstitute/2019_scWorkship" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="14 Functional Analysis | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  
  <meta name="twitter:description" content="This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook." />
  <meta name="twitter:image" content="https://broadinstitute.github.io/2019_scWorkshop/images/cover.png" />

<meta name="author" content="Orr Ashenberg">
<meta name="author" content="Dana Silverbush">
<meta name="author" content="Kirk Gosik">


<meta name="date" content="2019-02-25">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="correcting-batch-effects.html">
<link rel="next" href="pseudotime-cell-trajectories.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#course-overview"><i class="fa fa-check"></i><b>1.1</b> COURSE OVERVIEW</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#targeted-audience-assumed-background"><i class="fa fa-check"></i><b>1.2</b> TARGETED AUDIENCE &amp; ASSUMED BACKGROUND</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#course-format"><i class="fa fa-check"></i><b>1.3</b> COURSE FORMAT</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#getting-started"><i class="fa fa-check"></i><b>1.4</b> Getting Started</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#session-content"><i class="fa fa-check"></i><b>1.5</b> SESSION CONTENT</a><ul>
<li class="chapter" data-level="1.5.1" data-path="index.html"><a href="index.html#monday-classes-from-0930-to-1730-lunch-break-1-hr-40-min-of-total-coffee-breaks"><i class="fa fa-check"></i><b>1.5.1</b> Monday – Classes from 09:30 to 17:30 (lunch break-1 hr, 40 min of total coffee breaks)</a></li>
<li class="chapter" data-level="1.5.2" data-path="index.html"><a href="index.html#tuesday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.2</b> Tuesday – Classes from 09:30 to 17:30</a></li>
<li class="chapter" data-level="1.5.3" data-path="index.html"><a href="index.html#wednesday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.3</b> Wednesday – Classes from 09:30 to 17:30</a></li>
<li class="chapter" data-level="1.5.4" data-path="index.html"><a href="index.html#thursday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.4</b> Thursday – Classes from 09:30 to 17:30</a></li>
<li class="chapter" data-level="1.5.5" data-path="index.html"><a href="index.html#friday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.5</b> Friday – Classes from 09:30 to 17:30</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="scrna-seq-experimental-design.html"><a href="scrna-seq-experimental-design.html"><i class="fa fa-check"></i><b>2</b> scRNA-Seq Experimental Design</a></li>
<li class="chapter" data-level="3" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html"><i class="fa fa-check"></i><b>3</b> Understanding Sequencing Raw Data</a><ul>
<li class="chapter" data-level="3.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#class-environment"><i class="fa fa-check"></i><b>3.1</b> Class Environment</a><ul>
<li class="chapter" data-level="3.1.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#getting-into-aws-instance"><i class="fa fa-check"></i><b>3.1.1</b> Getting into AWS Instance</a></li>
<li class="chapter" data-level="3.1.2" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#rstudio"><i class="fa fa-check"></i><b>3.1.2</b> Rstudio</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#shell-and-unix-commands"><i class="fa fa-check"></i><b>3.2</b> Shell and Unix commands</a><ul>
<li class="chapter" data-level="3.2.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#common-linux-commands"><i class="fa fa-check"></i><b>3.2.1</b> Common Linux Commands</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#file-formats"><i class="fa fa-check"></i><b>3.3</b> File formats</a></li>
<li class="chapter" data-level="3.4" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#public-data-repositories"><i class="fa fa-check"></i><b>3.4</b> Public data repositories</a><ul>
<li class="chapter" data-level="3.4.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#cellranger10x"><i class="fa fa-check"></i><b>3.4.1</b> Cellranger/10x</a></li>
<li class="chapter" data-level="3.4.2" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#geo"><i class="fa fa-check"></i><b>3.4.2</b> GEO</a></li>
<li class="chapter" data-level="3.4.3" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#single-cell-portal"><i class="fa fa-check"></i><b>3.4.3</b> Single Cell Portal</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-preprocessing.html"><a href="data-preprocessing.html"><i class="fa fa-check"></i><b>4</b> Data Preprocessing</a></li>
<li class="chapter" data-level="5" data-path="processing-scrna-seq-data.html"><a href="processing-scrna-seq-data.html"><i class="fa fa-check"></i><b>5</b> Processing scRNA-Seq data</a></li>
<li class="chapter" data-level="6" data-path="transcriptome-quantification.html"><a href="transcriptome-quantification.html"><i class="fa fa-check"></i><b>6</b> Transcriptome Quantification</a></li>
<li class="chapter" data-level="7" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html"><i class="fa fa-check"></i><b>7</b> Introduction R/Bioconductor</a><ul>
<li class="chapter" data-level="7.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#installing-packages"><i class="fa fa-check"></i><b>7.1</b> Installing packages</a><ul>
<li class="chapter" data-level="7.1.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#cran"><i class="fa fa-check"></i><b>7.1.1</b> CRAN</a></li>
<li class="chapter" data-level="7.1.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#github"><i class="fa fa-check"></i><b>7.1.2</b> Github</a></li>
<li class="chapter" data-level="7.1.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#bioconductor"><i class="fa fa-check"></i><b>7.1.3</b> Bioconductor</a></li>
<li class="chapter" data-level="7.1.4" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#source"><i class="fa fa-check"></i><b>7.1.4</b> Source</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#installation-instructions"><i class="fa fa-check"></i><b>7.2</b> Installation instructions:</a></li>
<li class="chapter" data-level="7.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#data-typesclasses"><i class="fa fa-check"></i><b>7.3</b> Data-types/classes</a><ul>
<li class="chapter" data-level="7.3.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#numeric"><i class="fa fa-check"></i><b>7.3.1</b> Numeric</a></li>
<li class="chapter" data-level="7.3.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#characterstring"><i class="fa fa-check"></i><b>7.3.2</b> Character/String</a></li>
<li class="chapter" data-level="7.3.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#logical"><i class="fa fa-check"></i><b>7.3.3</b> Logical</a></li>
<li class="chapter" data-level="7.3.4" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#factors"><i class="fa fa-check"></i><b>7.3.4</b> Factors</a></li>
<li class="chapter" data-level="7.3.5" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#checking-classtype"><i class="fa fa-check"></i><b>7.3.5</b> Checking class/type</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#basic-data-structures"><i class="fa fa-check"></i><b>7.4</b> Basic data structures</a></li>
<li class="chapter" data-level="7.5" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#more-information"><i class="fa fa-check"></i><b>7.5</b> More information</a></li>
<li class="chapter" data-level="7.6" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#bioconductor-s4-classes"><i class="fa fa-check"></i><b>7.6</b> Bioconductor S4 Classes</a><ul>
<li class="chapter" data-level="7.6.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#s4-basics"><i class="fa fa-check"></i><b>7.6.1</b> S4 Basics</a></li>
<li class="chapter" data-level="7.6.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#s4-case-studies"><i class="fa fa-check"></i><b>7.6.2</b> S4 Case Studies</a></li>
<li class="chapter" data-level="7.6.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#s4-exercises"><i class="fa fa-check"></i><b>7.6.3</b> S4 Exercises</a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#grammer-of-graphics-ggplot2"><i class="fa fa-check"></i><b>7.7</b> Grammer of Graphics (ggplot2)</a><ul>
<li class="chapter" data-level="7.7.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#what-is-ggplot2"><i class="fa fa-check"></i><b>7.7.1</b> What is ggplot2?</a></li>
<li class="chapter" data-level="7.7.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#principles-of-ggplot2"><i class="fa fa-check"></i><b>7.7.2</b> Principles of ggplot2</a></li>
<li class="chapter" data-level="7.7.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#using-the-aes-mapping-function"><i class="fa fa-check"></i><b>7.7.3</b> Using the <code>aes</code> mapping function</a></li>
<li class="chapter" data-level="7.7.4" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#geoms"><i class="fa fa-check"></i><b>7.7.4</b> Geoms</a></li>
<li class="chapter" data-level="7.7.5" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#plotting-data-from-more-than-2-cells"><i class="fa fa-check"></i><b>7.7.5</b> Plotting data from more than 2 cells</a></li>
<li class="chapter" data-level="7.7.6" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#plotting-heatmaps"><i class="fa fa-check"></i><b>7.7.6</b> Plotting heatmaps</a></li>
<li class="chapter" data-level="7.7.7" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#principal-component-analysis"><i class="fa fa-check"></i><b>7.7.7</b> Principal Component Analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="expression-qc-and-normalization.html"><a href="expression-qc-and-normalization.html"><i class="fa fa-check"></i><b>8</b> Expression QC and Normalization</a></li>
<li class="chapter" data-level="9" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html"><i class="fa fa-check"></i><b>9</b> Data Wrangling scRNAseq</a></li>
<li class="chapter" data-level="10" data-path="identifying-cell-populations.html"><a href="identifying-cell-populations.html"><i class="fa fa-check"></i><b>10</b> Identifying Cell Populations</a><ul>
<li class="chapter" data-level="10.1" data-path="identifying-cell-populations.html"><a href="identifying-cell-populations.html#google-slides"><i class="fa fa-check"></i><b>10.1</b> Google Slides</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html"><i class="fa fa-check"></i><b>11</b> Feature Selection and Cluster Analysis</a><ul>
<li class="chapter" data-level="11.1" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#abstract"><i class="fa fa-check"></i><b>11.1</b> Abstract</a></li>
<li class="chapter" data-level="11.2" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#seurat-tutorial"><i class="fa fa-check"></i><b>11.2</b> Seurat Tutorial</a></li>
<li class="chapter" data-level="11.3" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#feature-selection"><i class="fa fa-check"></i><b>11.3</b> Feature Selection</a><ul>
<li class="chapter" data-level="11.3.1" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#differential-expression-variable-genes"><i class="fa fa-check"></i><b>11.3.1</b> Differential Expression / Variable Genes</a></li>
<li class="chapter" data-level="11.3.2" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#dimensionality-reduction"><i class="fa fa-check"></i><b>11.3.2</b> Dimensionality Reduction</a></li>
<li class="chapter" data-level="11.3.3" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#independent-components-analysis-ica"><i class="fa fa-check"></i><b>11.3.3</b> Independent Components Analysis (ICA)</a></li>
<li class="chapter" data-level="11.3.4" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#clustering"><i class="fa fa-check"></i><b>11.3.4</b> Clustering</a></li>
<li class="chapter" data-level="11.3.5" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#check-clusters"><i class="fa fa-check"></i><b>11.3.5</b> Check Clusters</a></li>
<li class="chapter" data-level="11.3.6" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#visualizingembedding"><i class="fa fa-check"></i><b>11.3.6</b> Visualizing/Embedding</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="12" data-path="batch-effects.html"><a href="batch-effects.html"><i class="fa fa-check"></i><b>12</b> Batch Effects</a></li>
<li class="chapter" data-level="13" data-path="correcting-batch-effects.html"><a href="correcting-batch-effects.html"><i class="fa fa-check"></i><b>13</b> Correcting Batch Effects</a></li>
<li class="chapter" data-level="14" data-path="functional-analysis.html"><a href="functional-analysis.html"><i class="fa fa-check"></i><b>14</b> Functional Analysis</a><ul>
<li class="chapter" data-level="14.1" data-path="functional-analysis.html"><a href="functional-analysis.html#google-slides-1"><i class="fa fa-check"></i><b>14.1</b> Google Slides</a></li>
<li class="chapter" data-level="14.2" data-path="functional-analysis.html"><a href="functional-analysis.html#gene-sets-and-signatures"><i class="fa fa-check"></i><b>14.2</b> Gene sets and signatures</a><ul>
<li class="chapter" data-level="14.2.1" data-path="functional-analysis.html"><a href="functional-analysis.html#cell-cycle"><i class="fa fa-check"></i><b>14.2.1</b> Cell Cycle</a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="functional-analysis.html"><a href="functional-analysis.html#pathway-analysis"><i class="fa fa-check"></i><b>14.3</b> Pathway analysis</a></li>
<li class="chapter" data-level="14.4" data-path="functional-analysis.html"><a href="functional-analysis.html#infercnv-honeybadger"><i class="fa fa-check"></i><b>14.4</b> inferCNV / honeybadger</a><ul>
<li class="chapter" data-level="14.4.1" data-path="functional-analysis.html"><a href="functional-analysis.html#create-the-infercnv-object"><i class="fa fa-check"></i><b>14.4.1</b> Create the InferCNV Object</a></li>
<li class="chapter" data-level="14.4.2" data-path="functional-analysis.html"><a href="functional-analysis.html#filtering-genes"><i class="fa fa-check"></i><b>14.4.2</b> Filtering genes</a></li>
<li class="chapter" data-level="14.4.3" data-path="functional-analysis.html"><a href="functional-analysis.html#normalize-each-cells-counts-for-sequencing-depth"><i class="fa fa-check"></i><b>14.4.3</b> Normalize each cell’s counts for sequencing depth</a></li>
<li class="chapter" data-level="14.4.4" data-path="functional-analysis.html"><a href="functional-analysis.html#perform-anscombe-normalization"><i class="fa fa-check"></i><b>14.4.4</b> Perform Anscombe normalization</a></li>
<li class="chapter" data-level="14.4.5" data-path="functional-analysis.html"><a href="functional-analysis.html#log-transform-the-normalized-counts"><i class="fa fa-check"></i><b>14.4.5</b> Log transform the normalized counts:</a></li>
<li class="chapter" data-level="14.4.6" data-path="functional-analysis.html"><a href="functional-analysis.html#apply-maximum-bounds-to-the-expression-data-to-reduce-outlier-effects"><i class="fa fa-check"></i><b>14.4.6</b> Apply maximum bounds to the expression data to reduce outlier effects</a></li>
<li class="chapter" data-level="14.4.7" data-path="functional-analysis.html"><a href="functional-analysis.html#initial-view-before-infercnv-operations"><i class="fa fa-check"></i><b>14.4.7</b> Initial view, before inferCNV operations:</a></li>
<li class="chapter" data-level="14.4.8" data-path="functional-analysis.html"><a href="functional-analysis.html#perform-smoothing-across-chromosomes"><i class="fa fa-check"></i><b>14.4.8</b> Perform smoothing across chromosomes</a></li>
<li class="chapter" data-level="14.4.9" data-path="functional-analysis.html"><a href="functional-analysis.html#subtract-the-reference-values-from-observations-now-have-logfold-change-values"><i class="fa fa-check"></i><b>14.4.9</b> Subtract the reference values from observations, now have log(fold change) values</a></li>
<li class="chapter" data-level="14.4.10" data-path="functional-analysis.html"><a href="functional-analysis.html#invert-log-values"><i class="fa fa-check"></i><b>14.4.10</b> Invert log values</a></li>
<li class="chapter" data-level="14.4.11" data-path="functional-analysis.html"><a href="functional-analysis.html#removing-noise"><i class="fa fa-check"></i><b>14.4.11</b> Removing noise</a></li>
<li class="chapter" data-level="14.4.12" data-path="functional-analysis.html"><a href="functional-analysis.html#remove-outlier-data-points"><i class="fa fa-check"></i><b>14.4.12</b> Remove outlier data points</a></li>
<li class="chapter" data-level="14.4.13" data-path="functional-analysis.html"><a href="functional-analysis.html#find-de-genes-by-comparing-the-mutant-types-to-normal-types-basic"><i class="fa fa-check"></i><b>14.4.13</b> Find DE genes by comparing the mutant types to normal types, BASIC</a></li>
<li class="chapter" data-level="14.4.14" data-path="functional-analysis.html"><a href="functional-analysis.html#additional-information"><i class="fa fa-check"></i><b>14.4.14</b> Additional Information</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="15" data-path="pseudotime-cell-trajectories.html"><a href="pseudotime-cell-trajectories.html"><i class="fa fa-check"></i><b>15</b> Pseudotime Cell Trajectories</a><ul>
<li class="chapter" data-level="15.1" data-path="pseudotime-cell-trajectories.html"><a href="pseudotime-cell-trajectories.html#google-slides-2"><i class="fa fa-check"></i><b>15.1</b> Google Slides</a></li>
<li class="chapter" data-level="15.2" data-path="pseudotime-cell-trajectories.html"><a href="pseudotime-cell-trajectories.html#comparison-abstract"><i class="fa fa-check"></i><b>15.2</b> Comparison Abstract</a></li>
<li class="chapter" data-level="15.3" data-path="pseudotime-cell-trajectories.html"><a href="pseudotime-cell-trajectories.html#what-is-pseudotime"><i class="fa fa-check"></i><b>15.3</b> What is Pseudotime</a></li>
<li class="chapter" data-level="15.4" data-path="pseudotime-cell-trajectories.html"><a href="pseudotime-cell-trajectories.html#inferring-pseudotime"><i class="fa fa-check"></i><b>15.4</b> Inferring Pseudotime</a><ul>
<li class="chapter" data-level="15.4.1" data-path="pseudotime-cell-trajectories.html"><a href="pseudotime-cell-trajectories.html#diffusion-maps"><i class="fa fa-check"></i><b>15.4.1</b> Diffusion Maps</a></li>
<li class="chapter" data-level="15.4.2" data-path="pseudotime-cell-trajectories.html"><a href="pseudotime-cell-trajectories.html#tree-methods"><i class="fa fa-check"></i><b>15.4.2</b> Tree Methods</a></li>
<li class="chapter" data-level="15.4.3" data-path="pseudotime-cell-trajectories.html"><a href="pseudotime-cell-trajectories.html#model-based-methods"><i class="fa fa-check"></i><b>15.4.3</b> Model Based Methods</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="16" data-path="spatial-genomics.html"><a href="spatial-genomics.html"><i class="fa fa-check"></i><b>16</b> Spatial Genomics</a></li>
<li class="chapter" data-level="17" data-path="functional-pseudotime-analysis.html"><a href="functional-pseudotime-analysis.html"><i class="fa fa-check"></i><b>17</b> Functional Pseudotime Analysis</a></li>
<li class="chapter" data-level="18" data-path="single-cell-multiomic-technologies.html"><a href="single-cell-multiomic-technologies.html"><i class="fa fa-check"></i><b>18</b> Single Cell Multiomic Technologies</a></li>
<li class="chapter" data-level="19" data-path="cite-seq-and-scatac-seq.html"><a href="cite-seq-and-scatac-seq.html"><i class="fa fa-check"></i><b>19</b> CITE-seq and scATAC-seq</a></li>
<li class="chapter" data-level="20" data-path="single-cell-resources.html"><a href="single-cell-resources.html"><i class="fa fa-check"></i><b>20</b> Single Cell Resources</a><ul>
<li class="chapter" data-level="20.1" data-path="single-cell-resources.html"><a href="single-cell-resources.html#comprehensive-list-of-single-cell-resources"><i class="fa fa-check"></i><b>20.1</b> Comprehensive list of single-cell resources</a></li>
<li class="chapter" data-level="20.2" data-path="single-cell-resources.html"><a href="single-cell-resources.html#computational-packages-for-single-cell-analysis"><i class="fa fa-check"></i><b>20.2</b> Computational packages for single-cell analysis</a></li>
<li class="chapter" data-level="20.3" data-path="single-cell-resources.html"><a href="single-cell-resources.html#elife-commentary-on-the-human-cell-atlas"><i class="fa fa-check"></i><b>20.3</b> eLife Commentary on the Human Cell Atlas</a></li>
<li class="chapter" data-level="20.4" data-path="single-cell-resources.html"><a href="single-cell-resources.html#online-courses"><i class="fa fa-check"></i><b>20.4</b> Online courses</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="functional-analysis" class="section level1">
<h1><span class="header-section-number">14</span> Functional Analysis</h1>
<div id="google-slides-1" class="section level2">
<h2><span class="header-section-number">14.1</span> Google Slides</h2>
<iframe src="https://docs.google.com/presentation/d/e/2PACX-1vT6PPJ9MLfG9Ri71Tn_BG0Ko8te2WPPf5pgeaMaopRwdfGqv3WkKWiuxXD86rXpR-5DSA62QvmOKJd4/embed?start=false&amp;loop=false&amp;delayms=3000" frameborder="0" width="760" height="569" allowfullscreen="true" mozallowfullscreen="true" webkitallowfullscreen="true">
</iframe>
</div>
<div id="gene-sets-and-signatures" class="section level2">
<h2><span class="header-section-number">14.2</span> Gene sets and signatures</h2>
<div id="cell-cycle" class="section level3">
<h3><span class="header-section-number">14.2.1</span> Cell Cycle</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">marrow &lt;-<span class="st"> </span><span class="kw">CellCycleScoring</span>(<span class="dt">object =</span> marrow, <span class="dt">s.genes =</span> s.genes, <span class="dt">g2m.genes =</span> g2m.genes, 
    <span class="dt">set.ident =</span> <span class="ot">TRUE</span>)

<span class="co"># view cell cycle scores and phase assignments</span>
<span class="kw">head</span>(<span class="dt">x =</span> marrow<span class="op">@</span>meta.data)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Visualize the distribution of cell cycle markers across</span>
<span class="kw">RidgePlot</span>(<span class="dt">object =</span> marrow, <span class="dt">features.plot =</span> <span class="kw">c</span>(<span class="st">&quot;PCNA&quot;</span>, <span class="st">&quot;TOP2A&quot;</span>, <span class="st">&quot;MCM6&quot;</span>, <span class="st">&quot;MKI67&quot;</span>), <span class="dt">nCol =</span> <span class="dv">2</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Running a PCA on cell cycle genes reveals, unsurprisingly, that cells</span>
<span class="co"># separate entirely by phase</span>
marrow &lt;-<span class="st"> </span><span class="kw">RunPCA</span>(<span class="dt">object =</span> marrow, <span class="dt">pc.genes =</span> <span class="kw">c</span>(s.genes, g2m.genes), <span class="dt">do.print =</span> <span class="ot">FALSE</span>)
<span class="kw">PCAPlot</span>(<span class="dt">object =</span> marrow)</code></pre></div>
</div>
</div>
<div id="pathway-analysis" class="section level2">
<h2><span class="header-section-number">14.3</span> Pathway analysis</h2>
</div>
<div id="infercnv-honeybadger" class="section level2">
<h2><span class="header-section-number">14.4</span> inferCNV / honeybadger</h2>
<p><a href="https://github.com/broadinstitute/inferCNV">Github Page</a></p>
<div id="create-the-infercnv-object" class="section level3">
<h3><span class="header-section-number">14.4.1</span> Create the InferCNV Object</h3>
<p>Reading in the raw counts matrix and meta data, populating the infercnv object</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">infercnv_obj =<span class="st"> </span><span class="kw">CreateInfercnvObject</span>(
  <span class="dt">raw_counts_matrix=</span><span class="st">&quot;../example/oligodendroglioma_expression_downsampled.counts.matrix&quot;</span>,
  <span class="dt">annotations_file=</span><span class="st">&quot;../example/oligodendroglioma_annotations_downsampled.txt&quot;</span>,
  <span class="dt">delim=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,
  <span class="dt">gene_order_file=</span><span class="st">&quot;../example/gencode_downsampled.txt&quot;</span>,
  <span class="dt">ref_group_names=</span><span class="kw">c</span>(<span class="st">&quot;Microglia/Macrophage&quot;</span>,<span class="st">&quot;Oligodendrocytes (non-malignant)&quot;</span>))</code></pre></div>
</div>
<div id="filtering-genes" class="section level3">
<h3><span class="header-section-number">14.4.2</span> Filtering genes</h3>
<p>Removing those genes that are very lowly expressed or present in very few cells</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># filter out low expressed genes</span>
cutoff=<span class="dv">1</span>
infercnv_obj &lt;-<span class="st"> </span><span class="kw">require_above_min_mean_expr_cutoff</span>(infercnv_obj, cutoff)
<span class="co"># filter out bad cells</span>
min_cells_per_gene=<span class="dv">3</span>
infercnv_obj &lt;-<span class="st"> </span><span class="kw">require_above_min_cells_ref</span>(infercnv_obj, <span class="dt">min_cells_per_gene=</span>min_cells_per_gene)
## for safe keeping
infercnv_orig_filtered =<span class="st"> </span>infercnv_obj
<span class="co">#plot_mean_chr_expr_lineplot(infercnv_obj)</span>
<span class="kw">save</span>(<span class="st">&#39;infercnv_obj&#39;</span>, <span class="dt">file =</span> <span class="st">&#39;../example_output/infercnv_obj.orig_filtered&#39;</span>)</code></pre></div>
</div>
<div id="normalize-each-cells-counts-for-sequencing-depth" class="section level3">
<h3><span class="header-section-number">14.4.3</span> Normalize each cell’s counts for sequencing depth</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">infercnv_obj &lt;-<span class="st"> </span>infercnv<span class="op">:::</span><span class="kw">normalize_counts_by_seq_depth</span>(infercnv_obj)</code></pre></div>
</div>
<div id="perform-anscombe-normalization" class="section level3">
<h3><span class="header-section-number">14.4.4</span> Perform Anscombe normalization</h3>
<p>Suggested by Matan for removing noisy variation at low counts</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">infercnv_obj &lt;-<span class="st"> </span>infercnv<span class="op">:::</span><span class="kw">anscombe_transform</span>(infercnv_obj)</code></pre></div>
<!--save('infercnv_obj', file='../example_output/infercnv_obj.anscombe')
```-->
</div>
<div id="log-transform-the-normalized-counts" class="section level3">
<h3><span class="header-section-number">14.4.5</span> Log transform the normalized counts:</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">infercnv_obj &lt;-<span class="st"> </span><span class="kw">log2xplus1</span>(infercnv_obj)</code></pre></div>
<!--save('infercnv_obj', file='../example_output/infercnv_obj.log_transformed')
```-->
</div>
<div id="apply-maximum-bounds-to-the-expression-data-to-reduce-outlier-effects" class="section level3">
<h3><span class="header-section-number">14.4.6</span> Apply maximum bounds to the expression data to reduce outlier effects</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">threshold =<span class="st"> </span><span class="kw">mean</span>(<span class="kw">abs</span>(<span class="kw">get_average_bounds</span>(infercnv_obj)))
infercnv_obj &lt;-<span class="st"> </span><span class="kw">apply_max_threshold_bounds</span>(infercnv_obj, <span class="dt">threshold=</span>threshold)</code></pre></div>
</div>
<div id="initial-view-before-infercnv-operations" class="section level3">
<h3><span class="header-section-number">14.4.7</span> Initial view, before inferCNV operations:</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot_cnv</span>(infercnv_obj,
         <span class="dt">out_dir=</span><span class="st">&#39;../example_output/&#39;</span>, 
         <span class="dt">output_filename=</span><span class="st">&#39;infercnv.logtransf&#39;</span>, 
         <span class="dt">x.range=</span><span class="st">&quot;auto&quot;</span>, 
         <span class="dt">title =</span> <span class="st">&quot;Before InferCNV (filtered &amp; log2 transformed)&quot;</span>, 
         <span class="dt">color_safe_pal =</span> <span class="ot">FALSE</span>, 
         <span class="dt">x.center =</span> <span class="kw">mean</span>(infercnv_obj<span class="op">@</span>expr.data))</code></pre></div>
</div>
<div id="perform-smoothing-across-chromosomes" class="section level3">
<h3><span class="header-section-number">14.4.8</span> Perform smoothing across chromosomes</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">infercnv_obj =<span class="st"> </span><span class="kw">smooth_by_chromosome</span>(infercnv_obj, <span class="dt">window_length=</span><span class="dv">101</span>, <span class="dt">smooth_ends=</span><span class="ot">TRUE</span>)</code></pre></div>
<!--# save('infercnv_obj', file='../example_output/infercnv_obj.smooth_by_chr')-->
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># re-center each cell</span>
infercnv_obj &lt;-<span class="st"> </span><span class="kw">center_cell_expr_across_chromosome</span>(infercnv_obj, <span class="dt">method =</span> <span class="st">&quot;median&quot;</span>)</code></pre></div>
<!--# save('infercnv_obj', file='../example_output/infercnv_obj.cells_recentered')-->
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot_cnv</span>(infercnv_obj, 
         <span class="dt">out_dir=</span><span class="st">&#39;../example_output/&#39;</span>,
         <span class="dt">output_filename=</span><span class="st">&#39;infercnv.chr_smoothed&#39;</span>, 
         <span class="dt">x.range=</span><span class="st">&quot;auto&quot;</span>, 
         <span class="dt">title =</span> <span class="st">&quot;chr smoothed and cells re-centered&quot;</span>, 
         <span class="dt">color_safe_pal =</span> <span class="ot">FALSE</span>)</code></pre></div>
</div>
<div id="subtract-the-reference-values-from-observations-now-have-logfold-change-values" class="section level3">
<h3><span class="header-section-number">14.4.9</span> Subtract the reference values from observations, now have log(fold change) values</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">infercnv_obj &lt;-<span class="st"> </span><span class="kw">subtract_ref_expr_from_obs</span>(infercnv_obj, <span class="dt">inv_log=</span><span class="ot">TRUE</span>)</code></pre></div>
<!--
save('infercnv_obj', file='../example_output/infercnv_obj.ref_subtracted')
```-->
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot_cnv</span>(infercnv_obj,
         <span class="dt">out_dir=</span><span class="st">&#39;../example_output/&#39;</span>,
         <span class="dt">output_filename=</span><span class="st">&#39;infercnv.ref_subtracted&#39;</span>, 
         <span class="dt">x.range=</span><span class="st">&quot;auto&quot;</span>, 
         <span class="dt">title=</span><span class="st">&quot;ref subtracted&quot;</span>, 
         <span class="dt">color_safe_pal =</span> <span class="ot">FALSE</span>)</code></pre></div>
</div>
<div id="invert-log-values" class="section level3">
<h3><span class="header-section-number">14.4.10</span> Invert log values</h3>
<p>Converting the log(FC) values to regular fold change values, centered at 1 (no fold change)</p>
<p>This is important because we want (1/2)x to be symmetrical to 1.5x, representing loss/gain of one chromosome region.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">infercnv_obj &lt;-<span class="st"> </span><span class="kw">invert_log2</span>(infercnv_obj)</code></pre></div>
<!--save('infercnv_obj', file='../example_output/infercnv_obj.inverted_log')
```-->
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot_cnv</span>(infercnv_obj, 
         <span class="dt">out_dir=</span><span class="st">&#39;../example_output/&#39;</span>,
         <span class="dt">output_filename=</span><span class="st">&#39;infercnv.inverted&#39;</span>, 
         <span class="dt">color_safe_pal =</span> <span class="ot">FALSE</span>, 
         <span class="dt">x.range=</span><span class="st">&quot;auto&quot;</span>, 
         <span class="dt">x.center=</span><span class="dv">1</span>, 
         <span class="dt">title =</span> <span class="st">&quot;inverted log FC to FC&quot;</span>)</code></pre></div>
</div>
<div id="removing-noise" class="section level3">
<h3><span class="header-section-number">14.4.11</span> Removing noise</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">infercnv_obj &lt;-<span class="st"> </span><span class="kw">clear_noise_via_ref_mean_sd</span>(infercnv_obj, <span class="dt">sd_amplifier =</span> <span class="fl">1.5</span>)</code></pre></div>
<!--save('infercnv_obj', file='../example_output/infercnv_obj.denoised')
```-->
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot_cnv</span>(infercnv_obj,
         <span class="dt">out_dir=</span><span class="st">&#39;../example_output/&#39;</span>,
         <span class="dt">output_filename=</span><span class="st">&#39;infercnv.denoised&#39;</span>, 
         <span class="dt">x.range=</span><span class="st">&quot;auto&quot;</span>, 
         <span class="dt">x.center=</span><span class="dv">1</span>, 
         <span class="dt">title=</span><span class="st">&quot;denoised&quot;</span>, 
         <span class="dt">color_safe_pal =</span> <span class="ot">FALSE</span>)</code></pre></div>
</div>
<div id="remove-outlier-data-points" class="section level3">
<h3><span class="header-section-number">14.4.12</span> Remove outlier data points</h3>
<p>This generally improves on the visualization</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">infercnv_obj =<span class="st"> </span><span class="kw">remove_outliers_norm</span>(infercnv_obj)</code></pre></div>
<!--save('infercnv_obj', file="../example_output/infercnv_obj.outliers_removed")
```-->
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="kw">plot_cnv</span>(infercnv_obj, 
         <span class="dt">out_dir=</span><span class="st">&#39;../example_output/&#39;</span>,
         <span class="dt">output_filename=</span><span class="st">&#39;infercnv.outliers_removed&#39;</span>, 
         <span class="dt">color_safe_pal =</span> <span class="ot">FALSE</span>, 
         <span class="dt">x.range=</span><span class="st">&quot;auto&quot;</span>, 
         <span class="dt">x.center=</span><span class="dv">1</span>, 
         <span class="dt">title =</span> <span class="st">&quot;outliers removed&quot;</span>)</code></pre></div>
</div>
<div id="find-de-genes-by-comparing-the-mutant-types-to-normal-types-basic" class="section level3">
<h3><span class="header-section-number">14.4.13</span> Find DE genes by comparing the mutant types to normal types, BASIC</h3>
<p>Runs a t-Test comparing tumor/normal for each patient and normal sample, and masks out those genes that are not significantly DE.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">plot_data =<span class="st"> </span>infercnv_obj<span class="op">@</span>expr.data
high_threshold =<span class="st"> </span><span class="kw">max</span>(<span class="kw">abs</span>(<span class="kw">quantile</span>(plot_data[plot_data <span class="op">!=</span><span class="st"> </span><span class="dv">0</span>], <span class="kw">c</span>(<span class="fl">0.05</span>, <span class="fl">0.95</span>))))  
low_threshold =<span class="st"> </span><span class="op">-</span><span class="dv">1</span> <span class="op">*</span><span class="st"> </span>high_threshold 
infercnv_obj2 &lt;-<span class="st"> </span>infercnv<span class="op">:::</span><span class="kw">mask_non_DE_genes_basic</span>(infercnv_obj, <span class="dt">test.use =</span> <span class="st">&#39;t&#39;</span>, <span class="dt">center_val=</span><span class="dv">1</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot_cnv</span>(infercnv_obj2, 
         <span class="dt">out_dir=</span><span class="st">&#39;../example_output/&#39;</span>,
         <span class="dt">output_filename=</span><span class="st">&#39;infercnv.non-DE-genes-masked&#39;</span>, 
         <span class="dt">color_safe_pal =</span> <span class="ot">FALSE</span>, 
         <span class="dt">x.range=</span><span class="kw">c</span>(low_threshold, high_threshold), 
         <span class="dt">x.center=</span><span class="dv">1</span>, 
         <span class="dt">title =</span> <span class="st">&quot;non-DE-genes-masked&quot;</span>)</code></pre></div>
</div>
<div id="additional-information" class="section level3">
<h3><span class="header-section-number">14.4.14</span> Additional Information</h3>
<div id="online-documentation" class="section level4">
<h4><span class="header-section-number">14.4.14.1</span> Online Documentation</h4>
<p>For additional explanations on files, usage, and a tutorial please visit the <a href="https://github.com/broadinstitute/inferCNV/wiki">wiki</a>.</p>
</div>
<div id="trinityctat" class="section level4">
<h4><span class="header-section-number">14.4.14.2</span> TrinityCTAT</h4>
<p>This tool is a part of the TrinityCTAT toolkit focused on leveraging the use of RNA-Seq to better understand cancer transcriptomes. To find out more please visit <a href="https://github.com/NCIP/Trinity_CTAT/wiki">TrinityCTAT</a></p>
</div>
<div id="applications" class="section level4">
<h4><span class="header-section-number">14.4.14.3</span> Applications</h4>
<p>This methodology was used in:</p>
<p><a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4123637/">Anoop P. Patel et al. Single-cell RNA-seq highlights intratumoral heterogeneity in primary glioblastoma. Science. 2014 Jun 20: 1396-1401</a></p>
<p><a href="http://www.ncbi.nlm.nih.gov/pubmed/27124452">Tirosh I et al.Dissecting the multicellular ecosystem of metastatic melanoma by single-cell RNA-seq. Science. 2016 Apr 8;352(6282):189-96</a></p>

</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="correcting-batch-effects.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="pseudotime-cell-trajectories.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["2019_scWorkshop.pdf", "2019_scWorkshop.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
